{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3ecaf4d4",
   "metadata": {},
   "source": [
    "# WTS sequencing data-processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "611bbbe9",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "62e1c94e",
   "metadata": {},
   "source": [
    "**Index**\n",
    "\n",
    "1. QC, trimming, and filtering\n",
    "   - Quality trimming (Trimmomatic)\n",
    "   - Re-check quality (FastQC; MultiQC)\n",
    "   - Remove rRNA with SortMeRNA\n",
    "   - Host filtering\n",
    "   - Check library sizes for each step\n",
    "2. Assembly\n",
    "   - Co-assembly\n",
    "   - Mini co-assemblies\n",
    "   - Per-sample assemblies\n",
    "   - Assessing assemblies\n",
    "3. Identify putative viral contigs\n",
    "   - VIBRANT\n",
    "   - VirSorter\n",
    "   - BLAST\n",
    "4. Summary table of putative viral contig methods\n",
    "5. Generate per-sample dereplicated fasta file of putative viral hits\n",
    "6. Quality assessment with checkV\n",
    "7. Dereplication of contigs across all samples\n",
    "8. Recreate summary table for dereplicated vOTUs\n",
    "   - BLAST alignments\n",
    "   - checkV\n",
    "   - Filtering of vOTUs\n",
    "9. Read mapping to obtain vOTU coverages\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbd70555",
   "metadata": {},
   "source": [
    "**Data**\n",
    "\n",
    "RNA transcriptomic data was obtained for 13 birds comprised of 9 faecal samples and 4 swab samples.\n",
    "\n",
    "Of the faecal samples:\n",
    "- S1-S7 are cloacitis birds Bravo, Alice, Taeatanga, Cyndy, Merv, Bella and Sinbad\n",
    "- S8 and S9 are control birds Nora and Scratch\n",
    "\n",
    "Of the swab samples:\n",
    "- S10 and S11 are cloacitis birds Uri and Hikoi\n",
    "- S12 and S13 are control birds Mukeke and Atareta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b41ba8e",
   "metadata": {},
   "source": [
    "## Trimmomatic\n",
    "\n",
    "Raw data concatenated in first step of WGS pipeline. RNA samples S1 - S13 moved to WTS folder under `0.Raw_concat`.\n",
    "\n",
    "**Trimmomatic: run**\n",
    "\n",
    "`sbatch scripts/wts_1_trimmomatic.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80d2376f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_1_trimmomatic\n",
    "#SBATCH --time 00:05:00\n",
    "#SBATCH --mem=12GB\n",
    "#SBATCH --array=1-13\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=10\n",
    "#SBATCH -e slurm_output/wts_1_trimmomatic_%a_%j.err\n",
    "#SBATCH -o slurm_output/wts_1_trimmomatic_%a_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load Trimmomatic/0.39-Java-1.8.0_144                \n",
    "\n",
    "# Make output directory \n",
    "mkdir -p 1.Qual_filtered_trimmomatic/\n",
    "\n",
    "# Set up variables for input path and output path\n",
    "inpath=0.Raw_concat\n",
    "outpath=1.Qual_filtered_trimmomatic\n",
    "\n",
    "# Make adapter file if not already created\n",
    "if [ ! -f iua.fna ]; then\n",
    "    echo \">FastQC_adapter\" > iua.fna\n",
    "    echo \"AGATCGGAAGAG\" >> iua.fna\n",
    "fi\n",
    "           \n",
    "# Quality filter and trim \n",
    "srun trimmomatic PE -threads 10 -phred33 \\\n",
    "${inpath}/S${SLURM_ARRAY_TASK_ID}_R1.fastq.gz ${inpath}/S${SLURM_ARRAY_TASK_ID}_R2.fastq.gz \\\n",
    "${outpath}/S${SLURM_ARRAY_TASK_ID}_R1.fastq S${SLURM_ARRAY_TASK_ID}_R1.single1.fastq \\\n",
    "${outpath}/S${SLURM_ARRAY_TASK_ID}_R2.fastq S${SLURM_ARRAY_TASK_ID}_R2.single2.fastq \\\n",
    "ILLUMINACLIP:iua.fna:1:25:7 CROP:115 SLIDINGWINDOW:4:30 MINLEN:50\n",
    "\n",
    "# Tidy up the singleton reads\n",
    "cat S${SLURM_ARRAY_TASK_ID}_R1.single1.fastq S${SLURM_ARRAY_TASK_ID}_R2.single2.fastq \\\n",
    "> ${outpath}/S${SLURM_ARRAY_TASK_ID}_single.fastq\n",
    "\n",
    "rm S${SLURM_ARRAY_TASK_ID}_R1.single1.fastq S${SLURM_ARRAY_TASK_ID}_R2.single2.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af3e79e7",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2c2be8f",
   "metadata": {},
   "source": [
    "## Fastqc analysis of post-trimmed reads\n",
    "\n",
    "`sbatch scripts/wts_1_qc_fastqc.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac402e62",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_1_qc_fastqc\n",
    "#SBATCH --time 01:00:00\n",
    "#SBATCH --mem 1GB\n",
    "#SBATCH --array=1-13\n",
    "#SBATCH --ntasks 1\n",
    "#SBATCH --cpus-per-task 2\n",
    "#SBATCH -e slurm_output/wts_1_qc_fastqc_%a_%j.err\n",
    "#SBATCH -o slurm_output/wts_1_qc_fastqc_%a_%j.out\n",
    "\n",
    "# Set up working directories\n",
    "cd <working_directory>/WTS/\n",
    "mkdir -p 1.Qual_filtered_trimmomatic/fastqc/\n",
    "\n",
    "# load modules\n",
    "module purge\n",
    "module load FastQC/0.11.9\n",
    "module load MultiQC/1.9-gimkl-2020a-Python-3.8.2\n",
    "\n",
    "# Run fastqc on each sample\n",
    "srun fastqc \\\n",
    "-o 1.Qual_filtered_trimmomatic/fastqc/ \\\n",
    "1.Qual_filtered_trimmomatic/S${SLURM_ARRAY_TASK_ID}_R1.fastq 1.Qual_filtered_trimmomatic/S${SLURM_ARRAY_TASK_ID}_R2.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63ab60ff",
   "metadata": {},
   "source": [
    "**MultiQC**\n",
    "\n",
    "`sbatch scripts/wts_1_qc_multiqc.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb483ddb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_1_qc_multiqc\n",
    "#SBATCH --time 00:10:00\n",
    "#SBATCH --mem 1GB\n",
    "#SBATCH --ntasks 1\n",
    "#SBATCH --cpus-per-task 2\n",
    "#SBATCH -e slurm_output/wts_1_qc_multiqc_%j.err\n",
    "#SBATCH -o slurm_output/wts_1_qc_multiqc_%j.out\n",
    "\n",
    "# Set up working directories\n",
    "cd <working_directory>/WTS/\n",
    "\n",
    "# load modules\n",
    "module purge\n",
    "module load FastQC/0.11.9\n",
    "module load MultiQC/1.9-gimkl-2020a-Python-3.8.2\n",
    "\n",
    "# Run multiqc to generate report for all samples\n",
    "srun multiqc -f \\\n",
    "-o 1.Qual_filtered_trimmomatic/fastqc/ \\\n",
    "1.Qual_filtered_trimmomatic/fastqc/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d080584e",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b1f8944",
   "metadata": {},
   "source": [
    "## SortMeRNA: filter out rRNA\n",
    "\n",
    "Remove ribosomal RNA from transcripts.\n",
    "\n",
    "**sortmerna v2.1 - Interleave paired files**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "825541d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change to working directory\n",
    "cd <working_directory>/WTS/\n",
    "\n",
    "# Load BBMap\n",
    "module purge\n",
    "module load BBMap/38.81-gimkl-2020a\n",
    "\n",
    "# Interleave paird files\n",
    "for i in {1..13}; do\n",
    "    reformat.sh \\\n",
    "    in=1.Qual_filtered_trimmomatic/S${i}_R#.fastq \\\n",
    "    out=1.Qual_filtered_trimmomatic/S${i}_interleaved.fastq\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc2ffd66",
   "metadata": {},
   "source": [
    "**SortMeRNA v2.1: Run**\n",
    "\n",
    "Paired data (interleaved files)\n",
    "\n",
    "`sbatch scripts/wts_1_sortmerna_interleaved.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "259426a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_1_sortmerna_interleaved\n",
    "#SBATCH --time 04:00:00\n",
    "#SBATCH --mem=2GB\n",
    "#SBATCH --array=1-13\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=32\n",
    "#SBATCH -e slurm_output/wts_1_sortmerna_interleaved_%a_%j.err\n",
    "#SBATCH -o slurm_output/wts_1_sortmerna_interleaved_%a_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/\n",
    "mkdir -p 1.rRNA_filtered/{aligned,unaligned}\n",
    "mkdir -p 1.rRNA_filtered/S${SLURM_ARRAY_TASK_ID}_interleaved\n",
    "\n",
    "cd 1.rRNA_filtered/S${SLURM_ARRAY_TASK_ID}_interleaved\n",
    "\n",
    "# Load modules\n",
    "module purge\n",
    "module load SortMeRNA/2.1-gimkl-2017a\n",
    "\n",
    "DATABASES_PATH='/nesi/project/uoa02469/Software/sortmerna-2.1'\n",
    "INPUT_DIR='<working_directory>'\n",
    "\n",
    "# Run sortmerna - paired files\n",
    "sortmerna \\\n",
    "--ref ${DATABASES_PATH}/rRNA_databases/silva-bac-16s-id90.fasta,${DATABASES_PATH}/index/silva-bac-16s-db:\\\n",
    "${DATABASES_PATH}/rRNA_databases/silva-bac-23s-id98.fasta,${DATABASES_PATH}/index/silva-bac-23s-db:\\\n",
    "${DATABASES_PATH}/rRNA_databases/silva-arc-16s-id95.fasta,${DATABASES_PATH}/index/silva-arc-16s-db:\\\n",
    "${DATABASES_PATH}/rRNA_databases/silva-arc-23s-id98.fasta,${DATABASES_PATH}/index/silva-arc-23s-db:\\\n",
    "${DATABASES_PATH}/rRNA_databases/silva-euk-18s-id95.fasta,${DATABASES_PATH}/index/silva-euk-18s-db:\\\n",
    "${DATABASES_PATH}/rRNA_databases/silva-euk-28s-id98.fasta,${DATABASES_PATH}/index/silva-euk-28s:\\\n",
    "${DATABASES_PATH}/rRNA_databases/rfam-5s-database-id98.fasta,${DATABASES_PATH}/index/rfam-5s-db:\\\n",
    "${DATABASES_PATH}/rRNA_databases/rfam-5.8s-database-id98.fasta,${DATABASES_PATH}/index/rfam-5.8s-db \\\n",
    "--reads $INPUT_DIR/1.Qual_filtered_trimmomatic/S${SLURM_ARRAY_TASK_ID}_interleaved.fastq \\\n",
    "--aligned $INPUT_DIR/1.rRNA_filtered/aligned/S${SLURM_ARRAY_TASK_ID}_paired_rRNA \\\n",
    "--other $INPUT_DIR/1.rRNA_filtered/unaligned/S${SLURM_ARRAY_TASK_ID}_paired_non_rRNA \\\n",
    "--paired_in --num_alignments 1 --fastx --log -v -a 32"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "857dd251",
   "metadata": {},
   "source": [
    "Single files (orphan reads output from trimmomatic)\n",
    "\n",
    "`sbatch scripts/wts_1_sortmerna_single.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "739059a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_1_sortmerna_single\n",
    "#SBATCH --time 00:45:00\n",
    "#SBATCH --mem=2GB\n",
    "#SBATCH --array=1-13\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=32\n",
    "#SBATCH -e slurm_output/wts_1_sortmerna_single_%a_%j.err\n",
    "#SBATCH -o slurm_output/wts_1_sortmerna_single_%a_%j.out\n",
    "\n",
    "# Change to working directory\n",
    "cd <working_directory>/WTS/\n",
    "mkdir -p 1.rRNA_filtered/aligned\n",
    "mkdir -p 1.rRNA_filtered/unaligned\n",
    "mkdir -p 1.rRNA_filtered/S${SLURM_ARRAY_TASK_ID}_single\n",
    "\n",
    "cd 1.rRNA_filtered/S${SLURM_ARRAY_TASK_ID}_single\n",
    "\n",
    "# Load modules\n",
    "module purge\n",
    "module load SortMeRNA/2.1-gimkl-2017a\n",
    "\n",
    "DATABASES_PATH='/nesi/project/uoa02469/Software/sortmerna-2.1'\n",
    "INPUT_DIR='<working_directory>'\n",
    "\n",
    "# Run sortmerna - single files\n",
    "sortmerna \\\n",
    "--ref ${DATABASES_PATH}/rRNA_databases/silva-bac-16s-id90.fasta,${DATABASES_PATH}/index/silva-bac-16s-db:\\\n",
    "${DATABASES_PATH}/rRNA_databases/silva-bac-23s-id98.fasta,${DATABASES_PATH}/index/silva-bac-23s-db:\\\n",
    "${DATABASES_PATH}/rRNA_databases/silva-arc-16s-id95.fasta,${DATABASES_PATH}/index/silva-arc-16s-db:\\\n",
    "${DATABASES_PATH}/rRNA_databases/silva-arc-23s-id98.fasta,${DATABASES_PATH}/index/silva-arc-23s-db:\\\n",
    "${DATABASES_PATH}/rRNA_databases/silva-euk-18s-id95.fasta,${DATABASES_PATH}/index/silva-euk-18s-db:\\\n",
    "${DATABASES_PATH}/rRNA_databases/silva-euk-28s-id98.fasta,${DATABASES_PATH}/index/silva-euk-28s:\\\n",
    "${DATABASES_PATH}/rRNA_databases/rfam-5s-database-id98.fasta,${DATABASES_PATH}/index/rfam-5s-db:\\\n",
    "${DATABASES_PATH}/rRNA_databases/rfam-5.8s-database-id98.fasta,${DATABASES_PATH}/index/rfam-5.8s-db \\\n",
    "--reads $INPUT_DIR/1.Qual_filtered_trimmomatic/S${SLURM_ARRAY_TASK_ID}_single.fastq \\\n",
    "--aligned $INPUT_DIR/1.rRNA_filtered/aligned/S${SLURM_ARRAY_TASK_ID}_single_rRNA \\\n",
    "--other $INPUT_DIR/1.rRNA_filtered/unaligned/S${SLURM_ARRAY_TASK_ID}_single_non_rRNA \\\n",
    "--paired_in --num_alignments 1 --fastx --log -v -a 32"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59f8b7bc",
   "metadata": {},
   "source": [
    "**SortmeRNA: Remove introduced empty lines**\n",
    "\n",
    "NOTE: sortmeRNA can sometimes introduce blank lines into the middle of files. This creates errors in the downstream files after deinterleaving (deinterleaving quits at the empty lines and the downstream files exclude all other reads in the file).\n",
    "\n",
    "Delete empty lines (or those with only whitespace) from the files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0cb9a5eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd <working_directory>/WTS/\n",
    "\n",
    "# \"Paired\" files\n",
    "for file in 1.rRNA_filtered/unaligned/*paired_non_rRNA.fastq; do\n",
    "    filename=$(basename ${file} .fastq)\n",
    "    echo ${filename}\n",
    "    grep '\\S' ${file} > 1.rRNA_filtered/unaligned/${filename}_filt.fastq\n",
    "done\n",
    "\n",
    "# \"Single\" files\n",
    "for file in 1.rRNA_filtered/unaligned/*single_non_rRNA.fastq; do\n",
    "    filename=$(basename ${file} .fastq)\n",
    "    echo ${filename}\n",
    "    grep '\\S' ${file} > 1.rRNA_filtered/unaligned/${filename}_filt.fastq\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d00afeb",
   "metadata": {},
   "source": [
    "**Deinterleave paired files**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f959f8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Change to working directory\n",
    "cd <working_directory>/WTS/\n",
    "\n",
    "# Load BBMap\n",
    "module purge\n",
    "module load BBMap/38.81-gimkl-2020a\n",
    "\n",
    "# Deinterleave paired files\n",
    "for i in {1..13}; do\n",
    "    reformat.sh \\\n",
    "    in=1.rRNA_filtered/unaligned/S${i}_paired_non_rRNA_filt.fastq \\\n",
    "    out1=1.rRNA_filtered/unaligned/S${i}_R1_non_rRNA.fastq \\\n",
    "    out2=1.rRNA_filtered/unaligned/S${i}_R2_non_rRNA.fastq\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46bb9d4c",
   "metadata": {},
   "source": [
    "**Downstream files**\n",
    "\n",
    "Filtered files for use downstream:\n",
    "- paired reads: `1.rRNA_filtered/unaligned/*_R1_non_rRNA.fastq`\n",
    "- single reads: `1.rRNA_filtered/unaligned/*single_non_rRNA_filt.fastq`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71e01651",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ea06d78",
   "metadata": {},
   "source": [
    "## Filter out host sequences\n",
    "\n",
    "**Preamble**\n",
    "\n",
    "Metagenome data derived from microbial communities associated with a host should ideally be filtered to remove any reads originating from host DNA. This may improve the quality and efficiency of downstream data processing, and is also an important consideration when working with metagenomes that may include data of a sensitive nature (and which may also need to be removed prior to making the data publicly available e.g. kākāpō). This is especially important for any studies involving human subjects or those involving samples derived from taonga species.\n",
    "\n",
    "There are several approaches that can be used to achieve this. The general principle is to map your reads to a reference genome (e.g. human genome) and remove those reads that map to the reference from the dataset.\n",
    "\n",
    "\n",
    "Here, we will map our reads against a masked kākāpō genome which is processed to hide sections that: are presumbed microbial contaminant in the reference; have high homology to microbial genes/genomes (e.g. ribosomes); or those that are of low complexity. This ensures that reads that would normally map to these sections of the kākāpō genome are not removed from the dataset (as genuine microbial reads that we wish to retain might also map to these regions), while all reads mapping to the rest of the genome are removed.\n",
    "\n",
    "The masked kākāpō genome was created in the WGS pipeline."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b05eaaed",
   "metadata": {},
   "source": [
    "JobID: 28538992\n",
    "Runtime: 5min\n",
    "Mem: 9GB"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93e8a423",
   "metadata": {},
   "source": [
    "**Host filtering: per-sample BBMap read mapping, slurm array**\n",
    "\n",
    "Note:\n",
    "- This step outputs fastq files where reads that map to the reference genome have been filtered out.\n",
    "- The output from `outu` is the filtered file for downstream use.\n",
    "- Host filtering here is run as a two step process for each sample: first, on the paired reads (R1 and R2), and then again for the unpaired (single) reads file.\n",
    "- The parameters are set based on the recomendations for host filtering outlined [here](https://www.seqanswers.com/forum/bioinformatics/bioinformatics-aa/37175-introducing-removehuman-human-contaminant-removal?t=42552)\n",
    "\n",
    "`sbatch scripts/wts_1_hostfilt_mapping.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "806af93c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_1_hostfilt_mapping\n",
    "#SBATCH --time 00:10:00\n",
    "#SBATCH --mem 28GB\n",
    "#SBATCH --ntasks 1\n",
    "#SBATCH --array=1-13\n",
    "#SBATCH --cpus-per-task 32\n",
    "#SBATCH -e slurm_output/wts_1_hostfilt_mapping_%a_%j.err\n",
    "#SBATCH -o slurm_output/wts_1_hostfilt_mapping_%a_%j.out\n",
    "\n",
    "# Set up working directories\n",
    "mkdir -p <working_directory>/WTS/1.host_filtered/\n",
    "cd <working_directory>/WTS/1.host_filtered/\n",
    "\n",
    "# Load BBMap module\n",
    "module purge\n",
    "module load BBMap/38.81-gimkl-2020a\n",
    "\n",
    "## Run bbmap\n",
    "\n",
    "# Copy over indexed masked kākāpō genome from WGS directory\n",
    "cp <working_directory>/WGS/1.host_filtered/<indexed_genome_files> <working_directory>/WTS/1.host_filtered/\n",
    "\n",
    "# Paired reads (R1 and R2)\n",
    "srun bbmap.sh -Xmx26g -t=32 \\\n",
    "minid=0.95 maxindel=3 bwr=0.16 bw=12 quickmatch fast minhits=2 qtrim=rl trimq=10 untrim \\\n",
    "in1=../1.rRNA_filtered/unaligned/S${SLURM_ARRAY_TASK_ID}_R1_non_rRNA.fastq \\\n",
    "in2=../1.rRNA_filtered/unaligned/S${SLURM_ARRAY_TASK_ID}_R2_non_rRNA.fastq \\\n",
    "outu1=S${SLURM_ARRAY_TASK_ID}_R1_hostFilt.fastq \\\n",
    "outu2=S${SLURM_ARRAY_TASK_ID}_R2_hostFilt.fastq\n",
    "\n",
    "# Unpaired (single) reads\n",
    "srun bbmap.sh -Xmx26g -t=32 \\\n",
    "minid=0.95 maxindel=3 bwr=0.16 bw=12 quickmatch fast minhits=2 qtrim=rl trimq=10 untrim \\\n",
    "in=../1.rRNA_filtered/unaligned/S${SLURM_ARRAY_TASK_ID}_single_non_rRNA_filt.fastq \\\n",
    "outu=S${SLURM_ARRAY_TASK_ID}_single_hostFilt.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e042fdd3",
   "metadata": {},
   "source": [
    "**Downstream files**\n",
    "\n",
    "Filtered files for use downstream:\n",
    "\n",
    "- paired reads: `1.host_filtered/*_R1_hostFilt.fastq; 1.host_filtered/*_R2_hostFilt.fastq`\n",
    "- single reads: `1.host_filtered/*single_hostFilt.fastq`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cabf8f4a",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "413f93c3",
   "metadata": {},
   "source": [
    "## Checking library sizes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb09c7b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd <working_directory>/WTS/\n",
    "\n",
    "# All raw files\n",
    "for file in 0.Raw_concat/*; do\n",
    "    echo ${file} \n",
    "done\n",
    "for file in 0.Raw_concat/*; do\n",
    "    zgrep -c '@' ${file} \n",
    "done\n",
    "\n",
    "# Trimmed files (concatenated parts 1 and 2 together)\n",
    "for file in 1.Qual_filtered_trimmomatic/*.fastq; do\n",
    "    echo ${file}\n",
    "done\n",
    "for file in 1.Qual_filtered_trimmomatic/*.fastq; do\n",
    "    grep -c '@' ${file} \n",
    "done\n",
    "\n",
    "# SortmeRNA-processed files\n",
    "for file in 1.rRNA_filtered/unaligned/*non_rRNA*; do\n",
    "    echo ${file}\n",
    "done\n",
    "for file in 1.rRNA_filtered/unaligned/*non_rRNA*; do\n",
    "    grep -c '@' ${file} \n",
    "done\n",
    "\n",
    "# Host filtered files\n",
    "for file in 1.host_filtered/*hostFilt*; do\n",
    "    echo ${file}\n",
    "done\n",
    "for file in 1.host_filtered/*hostFilt*; do\n",
    "    grep -c '@' ${file} \n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d2117560",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0509b50a",
   "metadata": {},
   "source": [
    "## Downstream processing options\n",
    "\n",
    "The following analyses regard virus detection from the RNA metatranscriptome data. Gene transcription profiles are created in the WGS pipeline."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25c55955",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4b736ec",
   "metadata": {},
   "source": [
    "## Assembly\n",
    "\n",
    "\n",
    "Assembly via SPAdes\n",
    "\n",
    "- rnaSPAdes is not actually suitable for assembling viral RNA (is designed more specifically for transcripts), while metaviralSPAdes is geared towards DNA viruses. Thus, here we use metaSPAdes for RNA viral contig assembly.\n",
    "\n",
    "**Concatenate reads for assembly**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a32ebf63",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd <working_directory>/WTS/\n",
    "\n",
    "mkdir -p 2.assembly_spades/infiles_concat\n",
    "\n",
    "cat 1.host_filtered/*_R1_hostFilt.fastq > 2.assembly_spades/infiles_concat/filtered_reads_R1.fastq\n",
    "cat 1.host_filtered/*_R2_hostFilt.fastq > 2.assembly_spades/infiles_concat/filtered_reads_R2.fastq\n",
    "cat 1.host_filtered/*_single_hostFilt.fastq > 2.assembly_spades/infiles_concat/filtered_reads_single.fastq\n",
    "\n",
    "# Mini co-assemblies\n",
    "## Healthy = S8, S9, S12, S13\n",
    "## Diseased = S1 - S7, S10, S11\n",
    "\n",
    "mkdir 1.host_filtered/Healthy\n",
    "mkdir 1.host_filtered/Diseased\n",
    "\n",
    "for i in {1:7,10,11}; do\n",
    "cp 1.host_filtered/S${i}*  1.host_filtered/Diseased/\n",
    "done\n",
    "\n",
    "rm 1.host_filtered/Diseased/S12*\n",
    "rm 1.host_filtered/Diseased/S13*\n",
    "\n",
    "for i in {8,9,12,13}; do\n",
    "cp 1.host_filtered/S${i}* 1.host_filtered/Healthy/\n",
    "done\n",
    "\n",
    "cat 1.host_filtered/Healthy/*_R1_hostFilt.fastq > 2.assembly_spades/infiles_concat/Healthy_reads_R1.fastq\n",
    "cat 1.host_filtered/Healthy/*_R2_hostFilt.fastq > 2.assembly_spades/infiles_concat/Healthy_reads_R2.fastq\n",
    "cat 1.host_filtered/Healthy/*_single_hostFilt.fastq > 2.assembly_spades/infiles_concat/Healthy_reads_single.fastq\n",
    "\n",
    "cat 1.host_filtered/Diseased/*_R1_hostFilt.fastq > 2.assembly_spades/infiles_concat/Diseased_reads_R1.fastq\n",
    "cat 1.host_filtered/Diseased/*_R2_hostFilt.fastq > 2.assembly_spades/infiles_concat/Diseased_reads_R2.fastq\n",
    "cat 1.host_filtered/Diseased/*_single_hostFilt.fastq > 2.assembly_spades/infiles_concat/Diseased_reads_single.fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53be1095",
   "metadata": {},
   "source": [
    "**Run co-assembly**\n",
    "\n",
    "NOTE: when changing the memory allocation, make sure to change it in both the SBATCH header and the actual spades call (the `-m` flag)\n",
    "\n",
    "`sbatch scripts/wts_2.coassembly_spades.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5051462d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_2.coassembly_spades\n",
    "#SBATCH --time 7:30:00\n",
    "#SBATCH --mem=40GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH -e slurm_output/wts_2.coassembly_spades_%j.err\n",
    "#SBATCH -o slurm_output/wts_2.coassembly_spades_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/\n",
    "mkdir -p 2.assembly_spades/coassembly\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load SPAdes/3.13.1-gimkl-2018b\n",
    "\n",
    "# Set up variables for input path and output path\n",
    "inpath=2.assembly_spades/infiles_concat\n",
    "outpath=2.assembly_spades\n",
    "\n",
    "# Run metaSPAdes\n",
    "srun spades.py --meta -t 16 -m 80 \\\n",
    "-1 2.assembly_spades/infiles_concat/filtered_reads_R1.fastq \\\n",
    "-2 2.assembly_spades/infiles_concat/filtered_reads_R2.fastq \\\n",
    "-s 2.assembly_spades/infiles_concat/filtered_reads_single.fastq \\\n",
    "-o 2.assembly_spades/coassembly/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5eae6bd",
   "metadata": {},
   "source": [
    "**Mini co-assemblies**\n",
    "\n",
    "`sbatch scripts/wts_2.healthy_coassembly_spades.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8649c6ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_2.healthy_coassembly_spades\n",
    "#SBATCH --time 3:30:00\n",
    "#SBATCH --mem=30GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH -e slurm_output/wts_2.healthy_coassembly_spades_%j.err\n",
    "#SBATCH -o slurm_output/wts_2.healthy_coassembly_spades_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/\n",
    "mkdir -p 2.assembly_spades/mini_co_assemblies/Healthy\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load SPAdes/3.13.1-gimkl-2018b\n",
    "\n",
    "# Set up variables for input path and output path\n",
    "inpath=2.assembly_spades/infiles_concat\n",
    "outpath=2.assembly_spades\n",
    "\n",
    "# Run metaSPAdes\n",
    "srun spades.py --meta -t 16 -m 50 \\\n",
    "-1 2.assembly_spades/infiles_concat/Healthy_reads_R1.fastq \\\n",
    "-2 2.assembly_spades/infiles_concat/Healthy_reads_R2.fastq \\\n",
    "-s 2.assembly_spades/infiles_concat/Healthy_reads_single.fastq \\\n",
    "-o 2.assembly_spades/mini_co_assemblies/Healthy/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c33fc14d",
   "metadata": {},
   "source": [
    "JobID: 28540090\n",
    "Runtime:1hr 10min\n",
    "Mem: 9GB"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef7aa2b4",
   "metadata": {},
   "source": [
    "`sbatch scripts/wts_2.diseased_coassembly_spades.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11a31615",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_2.diseased_coassembly_spades\n",
    "#SBATCH --time 2:30:00\n",
    "#SBATCH --mem=15GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH -e slurm_output/wts_2.diseased_coassembly_spades_%j.err\n",
    "#SBATCH -o slurm_output/wts_2.diseased_coassembly_spades_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/\n",
    "mkdir -p 2.assembly_spades/mini_co_assemblies/Diseased\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load SPAdes/3.13.1-gimkl-2018b\n",
    "\n",
    "# Set up variables for input path and output path\n",
    "inpath=2.assembly_spades/infiles_concat\n",
    "outpath=2.assembly_spades\n",
    "\n",
    "# Run metaSPAdes\n",
    "srun spades.py --meta -t 16 -m 50 \\\n",
    "-1 2.assembly_spades/infiles_concat/Diseased_reads_R1.fastq \\\n",
    "-2 2.assembly_spades/infiles_concat/Diseased_reads_R2.fastq \\\n",
    "-s 2.assembly_spades/infiles_concat/Diseased_reads_single.fastq \\\n",
    "-o 2.assembly_spades/mini_co_assemblies/Diseased/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f96a9e4",
   "metadata": {},
   "source": [
    "**Run individual assemblies as slurm array**\n",
    "\n",
    "`sbatch scripts/wts_2.assembly_spades.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1f3fc1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_2.assembly_spades\n",
    "#SBATCH --time 04:00:00\n",
    "#SBATCH --mem=10GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --array=1-13\n",
    "#SBATCH --cpus-per-task=6\n",
    "#SBATCH -e slurm_output/wts_2.assembly_spades_%a_%j.err\n",
    "#SBATCH -o slurm_output/wts_2.assembly_spades_%a_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/\n",
    "mkdir -p 2.assembly_spades/indv_assembly\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load SPAdes/3.13.1-gimkl-2018b\n",
    "\n",
    "# Run metaSPAdes\n",
    "srun spades.py --meta -t 6 -m 10 \\\n",
    "-1 1.host_filtered/S${SLURM_ARRAY_TASK_ID}_R1_hostFilt.fastq \\\n",
    "-2 1.host_filtered/S${SLURM_ARRAY_TASK_ID}_R2_hostFilt.fastq \\\n",
    "-s 1.host_filtered/S${SLURM_ARRAY_TASK_ID}_single_hostFilt.fastq \\\n",
    "-o 2.assembly_spades/indv_assembly/S${SLURM_ARRAY_TASK_ID}_assembly/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7671a64d",
   "metadata": {},
   "source": [
    "**Optional: Filtering out short contigs**\n",
    "    \n",
    "Note: Jemma Geoghegan recommended not filtering by length here for RNA viruses (as they can be quite short and/or you might still get useful information out of short fragments of viral genomes).\n",
    "\n",
    "If you wish to filter out short contigs, you can do so via `seqmagick`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1afab582",
   "metadata": {},
   "source": [
    "**Assessing assemblies**\n",
    "\n",
    "***Counting the number of contigs in each of the assemblies***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3631a3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd <working_directory>/WTS/\n",
    "\n",
    "## Count contigs\n",
    "# All contigs\n",
    "grep -c '>' 2.assembly_spades/coassembly/scaffolds.fasta\n",
    "grep -c '>' 2.assembly_spades/mini_coassembly/Healthy/scaffolds.fasta\n",
    "grep -c '>' 2.assembly_spades/mini_coassembly/Diseased/scaffolds.fasta\n",
    "\n",
    "for file in 2.assembly_spades/indv_assembly/*_indv_assembly/scaffolds.fasta; do\n",
    "    echo ${file}\n",
    "    grep -c '>' ${file} \n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e9289bb",
   "metadata": {},
   "source": [
    "**Assembly statistics via BBMap's stats.sh script**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "261c8ff8",
   "metadata": {},
   "outputs": [],
   "source": [
    "module purge\n",
    "module load BBMap/38.73-gimkl-2018b\n",
    "\n",
    "cd <working_directory>/WTS/\n",
    "\n",
    "# Run stats.sh\n",
    "stats.sh in=2.assembly_spades/coassembly/scaffolds.fasta\n",
    "stats.sh in=2.assembly_spades/mini_coassembly/Healthy/scaffolds.fasta\n",
    "stats.sh in=2.assembly_spades/mini_coassembly/Diseased/scaffolds.fasta\n",
    "\n",
    "for file in 2.assembly_spades/indv_assembly/*_indv_assembly/scaffolds.fasta; do\n",
    "    echo ${file}\n",
    "    stats.sh in=${file} \n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bbf58d6",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97dfac9a",
   "metadata": {},
   "source": [
    "## Identify putative viral contigs\n",
    "\n",
    "Below will cover:\n",
    "\n",
    "- VIBRANT\n",
    "- VirSorter2\n",
    "- Blast searchs (against NCBI-nr, NCBI-nt, custom RdRp database)\n",
    "    - Note: the blast searches require a modified version of the fasta file (remove line wrapping, and shorten sequence headers). You can use the script below for this, and then use the output modified files for all of the downstream processes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e877ece",
   "metadata": {},
   "source": [
    "### reformat_fa2oneline.py\n",
    "\n",
    "Python script created by Michael Hoggard to reformat fasta output files to be ready for inputting into blast (shorten contig headers and remove unneccesary info; remove linebreaks from within sequences).\n",
    "\n",
    "Example code:\n",
    "\n",
    "`reformat_fa2oneline.py \\`\n",
    "`-i 2.assembly/coassembly/scaffolds.fasta \\`\n",
    "`-o 3.virus_prediction/0.reformatted_fasta/scaffolds_reformated.fa`\n",
    "\n",
    "Output:\n",
    "\n",
    "`3.virus_prediction/0.reformatted_fasta/scaffolds_reformated.fa`: assembly fasta file reformatted for use downstream\n",
    "\n",
    "**reformat_fa2oneline.py: all assemblies**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa7c8559",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up working directories\n",
    "cd <working_directory>/WTS/\n",
    "mkdir -p 3.virus_prediction/0.reformatted_spades_fasta/\n",
    "\n",
    "# Load dependencies\n",
    "module purge\n",
    "module load Python/3.8.2-gimkl-2020a\n",
    "\n",
    "script_path='<working_directory>/WTS/3.virus_prediction'\n",
    "\n",
    "# run reformat_fa2oneline.py on co-assemblies\n",
    "3.virus_prediction/reformat_fa2oneline.py -i 2.assembly_spades/coassembly/scaffolds.fasta -o 3.virus_prediction/0.reformatted_spades_fasta/coassembly_scaffolds_reformated.fa\n",
    "3.virus_prediction/reformat_fa2oneline.py -i 2.assembly_spades/mini_coassembly/Healthy/scaffolds.fasta -o 3.virus_prediction/0.reformatted_spades_fasta/healthy_coassembly_scaffolds_reformated.fa\n",
    "3.virus_prediction/reformat_fa2oneline.py -i 2.assembly_spades/mini_coassembly/Diseased/scaffolds.fasta -o 3.virus_prediction/0.reformatted_spades_fasta/diseased_coassembly_scaffolds_reformated.fa\n",
    "\n",
    "# run on individual samples\n",
    "for i in {1..13}; do\n",
    "    ${script_path}/reformat_fa2oneline.py \\\n",
    "      -i 2.assembly_spades/indv_assembly/S${i}_indv_assembly/scaffolds.fasta \\\n",
    "      -o 3.virus_prediction/0.reformatted_spades_fasta/S${i}_scaffolds_reformated.fa\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc078cab",
   "metadata": {},
   "source": [
    "### VIBRANT\n",
    "\n",
    "See [VIBRANT github](https://github.com/AnantharamanLab/VIBRANT) for more information.\n",
    "\n",
    "NOTE: `$DB_PATH` variable is already set up with the VIBRANT module in NeSI\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "***VIBRANT: individual assemblies***\n",
    "\n",
    "`sbatch scripts/wts_3_vibrant.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "30df7204",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_vibrant\n",
    "#SBATCH --time 00:30:00\n",
    "#SBATCH --mem=6GB\n",
    "#SBATCH --array=1-13\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH -e slurm_output/wts_3_vibrant_%a_%j.err\n",
    "#SBATCH -o slurm_output/wts_3_vibrant_%a_%j.out\n",
    "\n",
    "# Set up working directories\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "mkdir -p 1.vibrant/individual_assemblies\n",
    "\n",
    "# Load dependencies\n",
    "module purge\n",
    "module load VIBRANT/1.2.1-gimkl-2020a\n",
    "\n",
    "# Run main analyses \n",
    "VIBRANT_run.py -t 16 \\\n",
    "-i 0.reformatted_spades_fasta/S${SLURM_ARRAY_TASK_ID}_scaffolds_reformated.fa \\\n",
    "-d $DB_PATH \\\n",
    "-folder 1.vibrant/individual_assemblies"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ec7da42",
   "metadata": {},
   "source": [
    "***VIBRANT: Co-assembly***\n",
    "\n",
    "`sbatch scripts/wts_3_vibrant_coassembly.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9aa993e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_vibrant_coassembly\n",
    "#SBATCH --time 00:30:00\n",
    "#SBATCH --mem=6GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH -e slurm_output/wts_3_vibrant_coassembly_%j.err\n",
    "#SBATCH -o slurm_output/wts_3_vibrant_coassembly_%j.out\n",
    "\n",
    "# Set up working directories\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "mkdir -p 1.vibrant/coassembly  \n",
    "\n",
    "# Load dependencies\n",
    "module purge\n",
    "module load VIBRANT/1.2.1-gimkl-2020a\n",
    "\n",
    "# Run main analyses \n",
    "VIBRANT_run.py -t 16 \\\n",
    "-i 0.reformatted_spades_fasta/coassembly_scaffolds_reformated.fa \\\n",
    "-d $DB_PATH \\\n",
    "-folder 1.vibrant/coassembly"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b167194d",
   "metadata": {},
   "source": [
    "***VIBRANT: Healthy co-assembly***\n",
    "\n",
    "`sbatch scripts/wts_3_vibrant_healthy.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7e90b2f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_vibrant_healthy\n",
    "#SBATCH --time 00:30:00\n",
    "#SBATCH --mem=4GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=8\n",
    "#SBATCH -e slurm_ouput/wts_3_vibrant_healthy_%j.err\n",
    "#SBATCH -o slurm_ouput/wts_3_vibrant_healthy_%j.out\n",
    "\n",
    "# Set up working directories\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "mkdir -p 1.vibrant/healthy \n",
    "\n",
    "# Load dependencies\n",
    "module purge\n",
    "module load VIBRANT/1.2.1-gimkl-2020a\n",
    "\n",
    "# Run main analyses \n",
    "VIBRANT_run.py -t 16 \\\n",
    "-i 0.reformatted_spades_fasta/healthy_coassembly_scaffolds_reformated.fa \\\n",
    "-d $DB_PATH \\\n",
    "-folder 1.vibrant/healthy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca22c389",
   "metadata": {},
   "source": [
    "***VIBRANT: Diseased co-assembly***\n",
    "\n",
    "`sbatch scripts/wts_3_vibrant_diseased.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78b98560",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_vibrant_diseased\n",
    "#SBATCH --time 00:30:00\n",
    "#SBATCH --mem=5GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=8\n",
    "#SBATCH -e slurm_ouput/wts_3_vibrant_diseased_%j.err\n",
    "#SBATCH -o slurm_ouput/wts_3_vibrant_diseased_%j.out\n",
    "\n",
    "# Set up working directories\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "mkdir -p 1.vibrant/diseased  \n",
    "\n",
    "# Load dependencies\n",
    "module purge\n",
    "module load VIBRANT/1.2.1-gimkl-2020a\n",
    "\n",
    "# Run main analyses \n",
    "VIBRANT_run.py -t 16 \\\n",
    "-i 0.reformatted_spades_fasta/diseased_coassembly_scaffolds_reformated.fa \\\n",
    "-d $DB_PATH \\\n",
    "-folder 1.vibrant/diseased"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "722c31cb",
   "metadata": {},
   "source": [
    "### VirSorter2\n",
    "\n",
    "See [VirSorter2 github](https://github.com/jiarong/VirSorter2) for more information and for a link to a tutorial/SoP, including recomended score threshold settings\n",
    "\n",
    "Running VirSorter2: parameters\n",
    "Key options:\n",
    "\n",
    "- `-d` database path (not part of the NeSI VirSorter2 module. Databases loaded locally from /nesi/project/uoa02469/Databases/virsorter2_20210909/)\n",
    "- `-i` input file (fa or fq format)\n",
    "- `-j` threads (jobs)\n",
    "- `--tmpdir` directory for temp files\n",
    "- `--rm-tmpdir` remove intermediate/temp files\n",
    "- `-l` label (prefix for output files; useful when re-running classify with different filtering)\n",
    "- `-w` working (output) directory\n",
    "\n",
    "\n",
    "n.b. to re-run with different filter threshold, just run 'classify' section, i.e.\n",
    "\n",
    "`virsorter run [options] classify`\n",
    "\n",
    "Filtering options include:\n",
    "\n",
    "- `--min-score FLOAT` min score to identify as viral (default 0.5)\n",
    "- `--min-length INTEGER` min seq length (default 0)\n",
    "- `--high-confidence-only` only output high confidence viral sequences; this is equivalent to screening final-viral-score.tsv with the following criteria: (max_score >= 0.9) OR (max_score >=0.7 AND hallmark >= 1) (default: False)\n",
    "- `--hallmark-required-on-short` require hallmark gene on short viral seqs (default 'short' length < 3kb) (default False)\n",
    "\n",
    "NOTE: `module unload XALT` also needs to be included after `module purge` to avoid a conflict when running the current VirSorter2 module in NeSI"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e34c7e8e",
   "metadata": {},
   "source": [
    "***VirSorter2: individual assemblies***\n",
    "\n",
    "`sbatch scripts/wts_3_vsort2.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4f142f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_vsort2\n",
    "#SBATCH --time 08:00:00\n",
    "#SBATCH --mem=8GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --array=1-13\n",
    "#SBATCH --cpus-per-task=32\n",
    "#SBATCH -e slurm_ouput/wts_3_vsort2_%a_%j.err\n",
    "#SBATCH -o slurm_ouput/wts_3_vsort2_%a_%J.out\n",
    "\n",
    "# Set up working directories\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "mkdir -p 2.virsorter2/individual_assemblies\n",
    "\n",
    "# Load module\n",
    "module purge\n",
    "module unload XALT\n",
    "module load VirSorter/2.2.3-gimkl-2020a-Python-3.8.2\n",
    "\n",
    "# Run virsorter2 - need access to Handley lab NeSI database to run this specific script \n",
    "srun virsorter run -j 32 \\\n",
    "-d /nesi/project/uoa02469/Databases/virsorter2_20210909/ \\\n",
    "-i 0.reformatted_spades_fasta/S${SLURM_ARRAY_TASK_ID}.scaffolds_reformated.fa \\\n",
    "--min-score 0.5 --include-groups dsDNAphage,NCLDV,RNA,ssDNA,lavidaviridae \\\n",
    "-l S${SLURM_ARRAY_TASK_ID} \\\n",
    "-w 2.virsorter2/individual_assemblies/S${SLURM_ARRAY_TASK_ID} \\\n",
    "--tmpdir ${SLURM_JOB_ID}.tmp \\\n",
    "--rm-tmpdir \\\n",
    "all \\\n",
    "--config LOCAL_SCRATCH=${TMPDIR:-/tmp}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "baca2f0b",
   "metadata": {},
   "source": [
    "JobID: 23526354 \n",
    "Mem: 2GB\n",
    "Time: 1.5 hr "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10b099d8",
   "metadata": {},
   "source": [
    "***VirSorter2: co-assembly***\n",
    "\n",
    "`sbatch scripts/wts_3_vsort2_coassembly.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf925364",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_vsort2_coassembly\n",
    "#SBATCH --time 08:00:00\n",
    "#SBATCH --mem=8GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=32\n",
    "#SBATCH -e slurm_ouput/wts_3_vsort2_coassembly_%j.err\n",
    "#SBATCH -o slurm_ouput/wts_3_vsort2_coassembly_%j.out\n",
    "\n",
    "# Set up working directories\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "mkdir -p 2.virsorter2/coassembly\n",
    "\n",
    "# Load module\n",
    "module purge\n",
    "module unload XALT\n",
    "module load VirSorter/2.2.3-gimkl-2020a-Python-3.8.2\n",
    " \n",
    "# Run virsorter2\n",
    "srun virsorter run -j 32 \\\n",
    "-d /nesi/project/uoa02469/Databases/virsorter2_20210909/ \\\n",
    "-i 0.reformatted_spades_fasta/coassembly_scaffolds_reformated.fa \\\n",
    "--min-score 0.5 --include-groups dsDNAphage,NCLDV,RNA,ssDNA,lavidaviridae \\\n",
    "-l coassembly \\\n",
    "-w 2.virsorter2/coassembly \\\n",
    "--tmpdir coassembly.tmp \\\n",
    "--rm-tmpdir \\\n",
    "all \\\n",
    "--config LOCAL_SCRATCH=${TMPDIR:-/tmp}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b94faf93",
   "metadata": {},
   "source": [
    "***VirSorter2: Healthy co-assembly***\n",
    "\n",
    "`sbatch scripts/wts_3_vsort2_healthy.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1df2ba7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_vsort2_healthy\n",
    "#SBATCH --time 03:00:00\n",
    "#SBATCH --mem=2GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH -e slurm_ouput/wts_3_vsort2_healthy_%j.err\n",
    "#SBATCH -o slurm_ouput/wts_3_vsort2_healthy_%j.out\n",
    "\n",
    "# Set up working directories\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "mkdir -p 2.virsorter2/healthy\n",
    "\n",
    "# Load module\n",
    "module purge\n",
    "module unload XALT\n",
    "module load VirSorter/2.2.3-gimkl-2020a-Python-3.8.2\n",
    " \n",
    "# Run virsorter2\n",
    "srun virsorter run -j 16 \\\n",
    "-d /nesi/project/uoa02469/Databases/virsorter2_20210909/ \\\n",
    "-i 0.reformatted_spades_fasta/healthy_coassembly_scaffolds_reformated.fa \\\n",
    "--min-score 0.5 --include-groups dsDNAphage,NCLDV,RNA,ssDNA,lavidaviridae \\\n",
    "-l healthy \\\n",
    "-w 2.virsorter2/healthy \\\n",
    "--tmpdir healthy.tmp \\\n",
    "--rm-tmpdir \\\n",
    "all \\\n",
    "--config LOCAL_SCRATCH=${TMPDIR:-/tmp}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9dda07ef",
   "metadata": {},
   "source": [
    "***VirSorter2: Diseased co-assembly***\n",
    "\n",
    "`sbatch scripts/wts_3_vsort2_diseased.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0ef2e3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_vsort2_diseased\n",
    "#SBATCH --time 04:00:00\n",
    "#SBATCH --mem=2GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=20\n",
    "#SBATCH -e slurm_ouput/wts_3_vsort2_diseased_%j.err\n",
    "#SBATCH -o slurm_ouput/wts_3_vsort2_diseased_%j.out\n",
    "\n",
    "# Set up working directories\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "mkdir -p 2.virsorter2/diseased\n",
    " \n",
    "# Load module\n",
    "module purge\n",
    "module unload XALT\n",
    "module load VirSorter/2.2.3-gimkl-2020a-Python-3.8.2\n",
    "\n",
    "# Run virsorter2\n",
    "srun virsorter run -j 20 \\\n",
    "-d /nesi/project/uoa02469/Databases/virsorter2_20210909/ \\\n",
    "-i 0.reformatted_spades_fasta/diseased_coassembly_scaffolds_reformated.fa \\\n",
    "--min-score 0.5 --include-groups dsDNAphage,NCLDV,RNA,ssDNA,lavidaviridae \\\n",
    "-l diseased \\\n",
    "-w 2.virsorter2/diseased \\\n",
    "--tmpdir diseased.tmp \\\n",
    "--rm-tmpdir \\\n",
    "all \\\n",
    "--config LOCAL_SCRATCH=${TMPDIR:-/tmp}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be9f1f7f",
   "metadata": {},
   "source": [
    "### BLAST searches\n",
    "\n",
    "Blast searches against NCBI nt, NCBI nr, and custom RdRp database (manually generated/curated by Jemma Geoghegan).\n",
    "\n",
    "Notes:\n",
    "\n",
    "- The nt database is loaded via `module load BLASTDB/2021-01` and works as is (and already includes taxonomy, so this can be called via e.g. `sskingdoms`)\n",
    "- The nr database was downloaded from NCBI (23/08/2021) and a dmnd database generated including taxonomy to allow for the inclusion of `sskingdoms`. This database is available in the Handley group databases directory: `/nesi/project/uoa02469/Databases/NCBI_20210116/nr/nr_wTax_FULL.dmnd`.\n",
    "- The RdRp database as custom generated by Jemma Geoghegan."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac4c5337",
   "metadata": {},
   "source": [
    "***BLAST search: nr (non-redundant protein database) sequences: individual assemblies***\n",
    "\n",
    "`sbatch scripts/wts_3_blast_nr.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4023c6e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_blast_nr\n",
    "#SBATCH --time 30:00:00\n",
    "#SBATCH --mem=36GB\n",
    "#SBATCH --array=1-13\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=24\n",
    "#SBATCH -e slurm_output/wts_3_blast_nr_%a_%j.err\n",
    "#SBATCH -o slurm_output/wts_3_blast_nr_%a_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "mkdir -p 3.blast/nr\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load DIAMOND/2.0.6-GCC-9.2.0\n",
    "\n",
    "# Database variable\n",
    "search_db='/nesi/project/uoa02469/Databases/NCBI_20210823_dmnd/nr_wTax_FULL.dmnd'\n",
    "\n",
    "# Run diamond blastx search\n",
    "srun diamond blastx \\\n",
    "-q 0.reformatted_fasta/S${SLURM_ARRAY_TASK_ID}.scaffolds_reformated.fa \\\n",
    "-d ${search_db} \\\n",
    "-o 3.blast/nr/S${SLURM_ARRAY_TASK_ID}_nr.txt \\\n",
    "-e 1E-5 -k 3 -p 24 -f 6 qseqid qlen sseqid stitle pident length evalue sskingdoms --more-sensitive"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bae54fb",
   "metadata": {},
   "source": [
    "***BLAST-nr: coassembly***\n",
    "\n",
    "`sbatch scripts/wts_3_blast_nr_coassembly`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b13ad2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_blast_nr_coassembly\n",
    "#SBATCH --time 30:00:00\n",
    "#SBATCH --mem=36GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=24\n",
    "#SBATCH -e slurm_output/wts_3_blast_nr_coassembly_%j.err\n",
    "#SBATCH -o slurm_output/wts_3_blast_nr_coassembly_%j.out\n",
    "#SBATCH --profile=task\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load DIAMOND/2.0.6-GCC-9.2.0\n",
    "\n",
    "# Database variable\n",
    "search_db='/nesi/project/uoa02469/Databases/NCBI_20210823_dmnd/nr_wTax_FULL.dmnd'\n",
    "\n",
    "# Run diamond blastx search\n",
    "srun diamond blastx \\\n",
    "-q 0.reformatted_spades_fasta/coassembly_scaffolds_reformated.fa \\\n",
    "-d ${search_db} \\\n",
    "-o 3.blast/nr/coassembly_nr.txt \\\n",
    "-e 1E-5 -k 3 -p 24 -f 6 qseqid qlen sseqid stitle pident length evalue --more-sensitive"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8be6a339",
   "metadata": {},
   "source": [
    "***BLAST-nr: mini-coassemblies***\n",
    "\n",
    "`sbatch scripts/wts_3_blast_nr_mini`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ca78599",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_blast_nr_mini\n",
    "#SBATCH --time 30:00:00\n",
    "#SBATCH --mem=36GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=24\n",
    "#SBATCH -e slurm_output/wts_3_blast_nr_mini_%j.err\n",
    "#SBATCH -o slurm_output/wts_3_blast_nr_mini_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load DIAMOND/2.0.6-GCC-9.2.0\n",
    "\n",
    "# Database variable\n",
    "search_db='/nesi/project/uoa02469/Databases/NCBI_20210823_dmnd/nr_wTax_FULL.dmnd'\n",
    "\n",
    "# Run diamond blastx search\n",
    "for j in healthy diseased; do\n",
    "    srun diamond blastx \\\n",
    "    -q 0.reformatted_fasta/${j}_coassembly_scaffolds_reformated.fa \\\n",
    "    -d ${search_db} \\\n",
    "    -o 3.blast/nr/${j}_coassembly_nr.txt \\\n",
    "    -e 1E-5 -k 3 -p 24 -f 6 qseqid qlen sseqid stitle pident length evalue sskingdoms --more-sensitive\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19e6bd0a",
   "metadata": {},
   "source": [
    "***BLAST search: nt (nucleotide database) sequences: individual assemblies***\n",
    "\n",
    "`sbatch scripts/wts_3_blast_nt.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d6f15085",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_blast_nt\n",
    "#SBATCH --time 06:00:00\n",
    "#SBATCH --mem=20GB\n",
    "#SBATCH --array=1-13\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=24\n",
    "#SBATCH -e slurm_output/wts_3_blast_nt_%a_%j.err\n",
    "#SBATCH -o slurm_output/wts_3_blast_nt_%a_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "mkdir -p 3.blast/nt\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load BLAST/2.10.0-GCC-9.2.0\n",
    "module load BLASTDB/2021-01\n",
    "\n",
    "# Database variable\n",
    "search_db=nt\n",
    "\n",
    "# Run diamond blastx search\n",
    "srun blastn \\\n",
    "-query 0.reformatted_fasta/S${SLURM_ARRAY_TASK_ID}.scaffolds_reformated.fa \\\n",
    "-db ${search_db} \\\n",
    "-out 3.blast/nt/S${SLURM_ARRAY_TASK_ID}_nt.txt \\\n",
    "-max_target_seqs 5 -num_threads 24 -evalue 1E-10 \\\n",
    "-outfmt \"6 qseqid sacc salltitles pident length evalue sskingdoms\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "da063c7a",
   "metadata": {},
   "source": [
    "***BLAST-nt: coassmebly***\n",
    "\n",
    "`sbatch scripts/wts_3_blast_nt_coassembly.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d74f9317",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_blast_nt_coassembly\n",
    "#SBATCH --time 06:00:00\n",
    "#SBATCH --mem=10GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=24\n",
    "#SBATCH -e slurm_output/wts_3_blast_nt_coassembly_%j.err\n",
    "#SBATCH -o slurm_output/wts_3_blast_nt_coassembly_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load BLAST/2.10.0-GCC-9.2.0\n",
    "module load BLASTDB/2021-01\n",
    "\n",
    "# Database variable\n",
    "search_db=nt\n",
    "\n",
    "# Run diamond blastx search\n",
    "srun blastn \\\n",
    "-query 0.reformatted_fasta/coassembly_scaffolds_reformated.fa \\\n",
    "-db ${search_db} \\\n",
    "-out 3.blast/nt/coassembly_nt.txt \\\n",
    "-max_target_seqs 5 -num_threads 24 -evalue 1E-10 \\\n",
    "-outfmt \"6 qseqid sacc salltitles pident length evalue sskingdoms\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9dff0cc",
   "metadata": {},
   "source": [
    "***BLAST-nt: mini-coassmeblies***\n",
    "\n",
    "`sbatch scripts/wts_3_blast_nt_mini.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e7b4aa2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_blast_nt_mini\n",
    "#SBATCH --time 06:00:00\n",
    "#SBATCH --mem=12GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=24\n",
    "#SBATCH -e slurm_output/wts_3_blast_nt_mini_%j.err\n",
    "#SBATCH -o slurm_output/wts_3_blast_nt_mini_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load BLAST/2.10.0-GCC-9.2.0\n",
    "module load BLASTDB/2021-01\n",
    "\n",
    "# Database variable\n",
    "search_db=nt\n",
    "\n",
    "# Run diamond blastx search\n",
    "for j in healthy diseased; do\n",
    "    srun blastn \\\n",
    "    -query 0.reformatted_fasta/${j}_coassembly_scaffolds_reformated.fa \\\n",
    "    -db ${search_db} \\\n",
    "    -out 3.blast/nt/${j}_coassembly_nt.txt \\\n",
    "    -max_target_seqs 5 -num_threads 24 -evalue 1E-10 \\\n",
    "    -outfmt \"6 qseqid sacc salltitles pident length evalue sskingdoms\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b74d12a",
   "metadata": {},
   "source": [
    "***BLAST search: RdRp (polymerase) viral sequences: individual assemblies***\n",
    "\n",
    "Notes:\n",
    "This searches against a custom database of RdRp sequences generated by Jemma Geoghegan.\n",
    "A DIAMOND formatted version of this database is available in the Handley group directory: /nesi/project/uoa02469/Databases/viral_rdrp_protein_database/viral_rdrp.dmnd\n",
    "\n",
    "\n",
    "`sbatch scripts/wts_3_blast_rdrp.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b67e43c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_blast_rdrp\n",
    "#SBATCH --time 00:05:00\n",
    "#SBATCH --mem=4GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=12\n",
    "#SBATCH -e slurm_output/wts_3_blast_rdrp_%j.err\n",
    "#SBATCH -o slurm_output/wts_3_blast_rdrp_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "mkdir -p 3.blast/rdrp\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load DIAMOND/0.9.32-GCC-9.2.0\n",
    "\n",
    "# Database variable\n",
    "search_db='/nesi/project/uoa02469/Databases/viral_rdrp_protein_database/viral_rdrp.dmnd'\n",
    "\n",
    "# Run diamond blastx search\n",
    "for i in {1..13}; do\n",
    "    srun diamond blastx \\\n",
    "    -q 0.reformatted_fasta/S${i}.scaffolds_reformated.fa \\\n",
    "    -d ${search_db} \\\n",
    "    -o 3.blast/rdrp/S${i}_RdRp.txt \\\n",
    "    -e 1E-5 -k 3 -p 24 -f 6 qseqid qlen sseqid stitle pident length evalue --more-sensitive\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86843d97",
   "metadata": {},
   "source": [
    "***BLAST-RdRp: coassmebly***\n",
    "\n",
    "`sbatch scripts/wts_3_blast_rdrp_coassembly.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ace79e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_blast_rdrp_coassembly\n",
    "#SBATCH --time 00:05:00\n",
    "#SBATCH --mem=4GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=12\n",
    "#SBATCH -e slurm_output/wts_3_blast_rdrp_coassembly_%j.err\n",
    "#SBATCH -o slurm_output/wts_3_blast_rdrp_coassembly_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load DIAMOND/0.9.32-GCC-9.2.0\n",
    "\n",
    "# Database variable\n",
    "search_db='/nesi/project/uoa02469/Databases/viral_rdrp_protein_database/viral_rdrp.dmnd'\n",
    "\n",
    "# Run diamond blastx search\n",
    "srun diamond blastx \\\n",
    "-q 0.reformatted_fasta/coassembly_scaffolds_reformated.fa \\\n",
    "-d ${search_db} \\\n",
    "-o 3.blast/rdrp/coassembly_RdRp.txt \\\n",
    "-e 1E-5 -k 3 -p 24 -f 6 qseqid qlen sseqid stitle pident length evalue --more-sensitive\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "044905a5",
   "metadata": {},
   "source": [
    "***BLAST-RdRp: mini-coassmeblies***\n",
    "\n",
    "`sbatch scripts/wts_3_blast_rdrp_mini.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a55f4b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_3_blast_rdrp_mini\n",
    "#SBATCH --time 00:05:00\n",
    "#SBATCH --mem=4GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=12\n",
    "#SBATCH -e slurm_output/wts_3_blast_rdrp_mini_%j.err\n",
    "#SBATCH -o slurm_output/wts_3_blast_rdrp_mini_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/3.virus_prediction/\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load BLAST/2.10.0-GCC-9.2.0\n",
    "module load BLASTDB/2021-01\n",
    "\n",
    "# Database variable\n",
    "search_db=nt\n",
    "\n",
    "# Run diamond blastx search\n",
    "for j in healthy diseased; do\n",
    "    srun blastn \\\n",
    "    -query 0.reformatted_fasta/${j}_coassembly_scaffolds_reformated.fa \\\n",
    "    -db ${search_db} \\\n",
    "    -out 3.blast/rdrp/${j}_coassembly_RdRp.txt \\\n",
    "    -max_target_seqs 5 -num_threads 24 -evalue 1E-10 \\\n",
    "    -outfmt \"6 qseqid sacc salltitles pident length evalue sskingdoms\"\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea5c166d",
   "metadata": {},
   "source": [
    "JobID: 23540744"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d76ab8f5",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ec892f3",
   "metadata": {},
   "source": [
    "## Summary table of all methods\n",
    "\n",
    "- `summarise_viral_contigs.py` generated by Michael Hoggard\n",
    "- For each sample, generate summary table including putative virus results for each method\n",
    "\n",
    "Note:\n",
    "\n",
    "- In output summary table, individual contigs will generally be repeated over multiple lines due to multiple blast database hits for the contig.\n",
    "- `summary_table_VIRUSES.txt` output file is filtered to retain any hits made by VIBRANT, VirSorter2, RdRp database hits, or nt or nr hits with taxonomy matching viruses."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3ab7dd3",
   "metadata": {},
   "source": [
    "***summarise_viral_contigs.py: All samples¶***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "480f1466",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load module(s)\n",
    "module purge\n",
    "module load Python/3.8.2-gimkl-2020a\n",
    "\n",
    "# Script path\n",
    "script_path='/nesi/project/uoa02469/custom-scripts/MikeH'\n",
    "\n",
    "## Output filepath\n",
    "cd <working_directory>/WTS/\n",
    "mkdir -p 3.virus_prediction/4.summary_tables\n",
    "\n",
    "# Run script: individual samples\n",
    "for i in {1..13}; do\n",
    "    ${script_path}/summarise_viral_contigs.py \\\n",
    "    --blastn_nt 3.virus_prediction/3.blast/nt/S${i}_nt.txt \\\n",
    "    --blastx_nr 3.virus_prediction/3.blast/nr/S${i}_nr.txt \\\n",
    "    --blastx_rdrp 3.virus_prediction/3.blast/rdrp/S${i}_RdRp.txt \\\n",
    "    --vibrant 3.virus_prediction/1.vibrant/individual_assemblies/VIBRANT_S${i}.scaffolds_reformated/VIBRANT_results_S${i}.scaffolds_reformated/VIBRANT_summary_results_S${i}.scaffolds_reformated.tsv \\\n",
    "    --virsorter2 3.virus_prediction/2.virsorter2/individual_assemblies/S${i}/S${i}-final-viral-score.tsv \\\n",
    "    --out_prefix 3.virus_prediction/4.summary_tables/S${i}_summary_table\n",
    "done\n",
    "\n",
    "# Coassembly\n",
    "${script_path}/summarise_viral_contigs.py \\\n",
    "    --blastn_nt 3.virus_prediction/3.blast/nt/coassembly_nt.txt \\\n",
    "    --blastx_nr 3.virus_prediction/3.blast/nr/coassembly_nr.txt \\\n",
    "    --blastx_rdrp 3.virus_prediction/3.blast/rdrp/coassembly_RdRp.txt \\\n",
    "    --vibrant 3.virus_prediction/1.vibrant/coassembly/VIBRANT_coassembly.scaffolds_reformated/VIBRANT_results_coassembly.scaffolds_reformated/VIBRANT_summary_results_coassembly.scaffolds_reformated.tsv \\\n",
    "    --virsorter2 3.virus_prediction/2.virsorter2/coassembly/coassembly-final-viral-score.tsv \\\n",
    "    --out_prefix 3.virus_prediction/4.summary_tables/coassembly_summary_table\n",
    "\n",
    "# Mini assemblies\n",
    "${script_path}/summarise_viral_contigs.py \\\n",
    "    --blastn_nt 3.virus_prediction/3.blast/nt/healthy_coassembly_nt.txt \\\n",
    "    --blastx_nr 3.virus_prediction/3.blast/nr/healthy_coassembly_nr.txt \\\n",
    "    --blastx_rdrp 3.virus_prediction/3.blast/rdrp/healthy_coassembly_RdRp.txt \\\n",
    "    --vibrant 3.virus_prediction/1.vibrant/healthy/VIBRANT_healthy_coassembly.scaffolds_reformated/VIBRANT_results_healthy_coassembly.scaffolds_reformated/VIBRANT_summary_results_healthy_coassembly.scaffolds_reformated.tsv \\\n",
    "    --virsorter2 3.virus_prediction/2.virsorter2/healthy/healthy-final-viral-score.tsv \\\n",
    "    --out_prefix 3.virus_prediction/4.summary_tables/healthy_coassembly_summary_table\n",
    "\n",
    "${script_path}/summarise_viral_contigs.py \\\n",
    "    --blastn_nt 3.virus_prediction/3.blast/nt/diseased_coassembly_nt.txt \\\n",
    "    --blastx_nr 3.virus_prediction/3.blast/nr/diseased_coassembly_nr.txt \\\n",
    "    --blastx_rdrp 3.virus_prediction/3.blast/rdrp/diseased_coassembly_RdRp.txt \\\n",
    "    --vibrant 3.virus_prediction/1.vibrant/diseased/VIBRANT_diseased_coassembly.scaffolds_reformated/VIBRANT_results_diseased_coassembly.scaffolds_reformated/VIBRANT_summary_results_diseased_coassembly.scaffolds_reformated.tsv \\\n",
    "    --virsorter2 3.virus_prediction/2.virsorter2/diseased/diseased-final-viral-score.tsv \\\n",
    "    --out_prefix 3.virus_prediction/4.summary_tables/diseased_coassembly_summary_table"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd183d96",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03fe056e",
   "metadata": {},
   "source": [
    "## Generate per-sample dereplicated fasta file of putative viral hits\n",
    "\n",
    "`virome_per_sample_derep.py` generated by Michael Hoggard\n",
    "\n",
    "- Creates fasta file combining results from all tools (based on `summary_table_VIRUSES.txt` generated above)\n",
    "    - Reads in assembly fasta file and writes out any found in summary table contig_ID to sampleX_derep.fasta\n",
    "    - (If provided) Reads in VIBRANT fasta output file and appends any *prophage* (i.e. IDs containing 'fragment') found in summary table contig_ID to sampleX_derep.fasta\n",
    "    - (If provided) Reads in VirSorter2 fasta output file and appends any *prophage* (i.e. IDs containing 'partial') found in summary table contig_ID to sampleX_derep.fasta\n",
    "    - Filters out any included full contigs that were also identified as (excised) prophage via VIBRANT or VirSorter2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "282327b4",
   "metadata": {},
   "source": [
    "***Run virome_per_sample_derep.py***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49b73662",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load module(s)\n",
    "module purge\n",
    "module load Python/3.8.2-gimkl-2020a\n",
    "\n",
    "# Script path\n",
    "script_path='/nesi/project/uoa02469/custom-scripts/MikeH'\n",
    "\n",
    "# Directories\n",
    "cd <working_directory>/WTS/\n",
    "mkdir -p 3.virus_prediction/5.perSample_derep\n",
    "\n",
    "# Run for individual samples\n",
    "for i in {1..13}; do\n",
    "    ${script_path}/virome_per_sample_derep.py \\\n",
    "    --assembly_fasta 3.virus_prediction/0.reformatted_fasta/S${i}.scaffolds_reformated.fa \\\n",
    "    --summary_table 3.virus_prediction/4.summary_tables/S${i}_summary_table_VIRUSES.txt \\\n",
    "    --vibrant 3.virus_prediction/1.vibrant/VIBRANT_S${i}.scaffolds_reformated/VIBRANT_phages_S${i}.scaffolds_reformated/S${i}.scaffolds_reformated.phages_combined.fna \\\n",
    "    --virsorter2 3.virus_prediction/2.virsorter2/S${i}/S${i}-final-viral-combined.fa \\\n",
    "    --output 3.virus_prediction/5.perSample_derep/S${i}_perSample_derep.fasta\n",
    "done\n",
    "\n",
    "\n",
    "# Coassembly\n",
    "${script_path}/virome_per_sample_derep.py \\\n",
    "    --assembly_fasta 3.virus_prediction/0.reformatted_spades_fasta/coassembly_scaffolds_reformated.fa \\\n",
    "    --summary_table 3.virus_prediction/4.summary_tables/coassembly_summary_table_VIRUSES.txt \\\n",
    "    --vibrant 3.virus_prediction/1.vibrant/coassembly/VIBRANT_coassembly_scaffolds_reformated/VIBRANT_phages_coassembly_scaffolds_reformated/coassembly_scaffolds_reformated.phages_combined.fna \\\n",
    "    --virsorter2 3.virus_prediction/2.virsorter2/coassembly/coassembly-final-viral-combined.fa \\\n",
    "    --output 3.virus_prediction/5.perSample_derep/coassembly_perSample_derep.fasta\n",
    "\n",
    "# Mini assemblies\n",
    "${script_path}/virome_per_sample_derep.py \\\n",
    "    --assembly_fasta 3.virus_prediction/0.reformatted_spades_fasta/healthy_coassembly_scaffolds_reformated.fa \\\n",
    "    --summary_table 3.virus_prediction/4.summary_tables/healthy_coassembly_summary_table_VIRUSES.txt \\\n",
    "    --vibrant 3.virus_prediction/1.vibrant/healthy/VIBRANT_healthy_coassembly_scaffolds_reformated/VIBRANT_phages_healthy_coassembly_scaffolds_reformated/healthy_coassembly_scaffolds_reformated.phages_combined.fna \\\n",
    "    --virsorter2 3.virus_prediction/2.virsorter2/healthy/healthy-final-viral-combined.fa \\\n",
    "    --output 3.virus_prediction/5.perSample_derep/healthy_coassembly_perSample_derep.fasta\n",
    "\n",
    "${script_path}/virome_per_sample_derep.py \\\n",
    "    --assembly_fasta 3.virus_prediction/0.reformatted_spades_fasta/diseased_coassembly_scaffolds_reformated.fa \\\n",
    "    --summary_table 3.virus_prediction/4.summary_tables/diseased_coassembly_summary_table_VIRUSES.txt \\\n",
    "    --vibrant 3.virus_prediction/1.vibrant/diseased/VIBRANT_diseased_coassembly_scaffolds_reformated/VIBRANT_phages_diseased_coassembly_scaffolds_reformated/diseased_coassembly_scaffolds_reformated.phages_combined.fna \\\n",
    "    --virsorter2 3.virus_prediction/2.virsorter2/diseased/diseased-final-viral-combined.fa \\\n",
    "    --output 3.virus_prediction/5.perSample_derep/diseased_coassembly_perSample_derep.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "882ad55e",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdc9e3c9",
   "metadata": {},
   "source": [
    "## CheckV per sample"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ecc529b",
   "metadata": {},
   "source": [
    "More info about checkV [here](https://bitbucket.org/berkeleylab/checkv/src/master/)\n",
    "\n",
    "Run checkv on ouput from dereplication of contigs identified by all methods to obtain quality information.\n",
    "\n",
    "\n",
    "***CheckV: individual assemblies***\n",
    "\n",
    "`sbatch scripts/wts_4_checkv_perSample.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63dc5fc4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_4_checkv_perSample\n",
    "#SBATCH --time 00:10:00\n",
    "#SBATCH --mem=4GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --array=1-13\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH -e slurm_output/wts_4_checkv_perSample_%a_%j.err\n",
    "#SBATCH -o slurm_output/wts_4_checkv_perSample_%a_%j.out\n",
    "\n",
    "# Set up working directories\n",
    "cd <working_directory>/WTS/\n",
    "mkdir -p 3.virus_prediction/6.perSample_checkv\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load CheckV/0.7.0-gimkl-2020a-Python-3.8.2\n",
    "\n",
    "# Run checkv\n",
    "checkv_in=\"3.virus_prediction/5.perSample_derep/S${SLURM_ARRAY_TASK_ID}_perSample_derep.fasta\"\n",
    "checkv_out=\"3.virus_prediction/6.perSample_checkv/S${SLURM_ARRAY_TASK_ID}_perSample_derep_checkv_out\"\n",
    "srun checkv end_to_end ${checkv_in} ${checkv_out} -t 16 --quiet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0955abdc",
   "metadata": {},
   "source": [
    "***CheckV: coassemblies***\n",
    "\n",
    "`sbatch scripts/wts_4_checkv_perSample_coassemblies.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b27102bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_4_checkv_perSample\n",
    "#SBATCH --time 01:00:00\n",
    "#SBATCH --mem=4GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH -e slurm_output/wts_4_checkv_perSample_coassemblies_%j.err\n",
    "#SBATCH -o slurm_output/wts_4_checkv_perSample_coassemblies_%j.out\n",
    "\n",
    "# Set up working directories\n",
    "cd <working_directory>/WTS/\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load CheckV/0.7.0-gimkl-2020a-Python-3.8.2\n",
    "\n",
    "# Run checkv\n",
    "checkv_in=\"3.virus_prediction/5.perSample_derep/coassembly_perSample_derep.fasta\"\n",
    "checkv_out=\"3.virus_prediction/6.perSample_checkv/coassembly_perSample_derep_checkv_out\"\n",
    "srun checkv end_to_end ${checkv_in} ${checkv_out} -t 16 --quiet\n",
    "\n",
    "checkv_in=\"3.virus_prediction/5.perSample_derep/healthy_coassembly_perSample_derep.fasta\"\n",
    "checkv_out=\"3.virus_prediction/6.perSample_checkv/healthy_coassembly_perSample_derep_checkv_out\"\n",
    "srun checkv end_to_end ${checkv_in} ${checkv_out} -t 16 --quiet\n",
    "\n",
    "checkv_in=\"3.virus_prediction/5.perSample_derep/diseased_coassembly_perSample_derep.fasta\"\n",
    "checkv_out=\"3.virus_prediction/6.perSample_checkv/diseased_coassembly_perSample_derep_checkv_out\"\n",
    "srun checkv end_to_end ${checkv_in} ${checkv_out} -t 16 --quiet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a018508c",
   "metadata": {},
   "source": [
    "***CheckV perSample: Concatenate output fasta files for downstream use (viruses.fna and proviruses.fna)***\n",
    "\n",
    "Also modify prophage contig headers for downstream use"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "823a5999",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd <working_directory>/WTS/\n",
    "\n",
    "# Individual assemblies\n",
    "for i in {1..13}; do\n",
    "    # concatenate viruses and prophage files\n",
    "    cat 3.virus_prediction/6.perSample_checkv/S${i}_perSample_derep_checkv_out/viruses.fna 3.virus_prediction/6.perSample_checkv/S${i}_perSample_derep_checkv_out/proviruses.fna > 3.virus_prediction/6.perSample_checkv/S${i}_perSample_derep_checkv_out/cat_checkv_out.fasta \n",
    "    # modify checkv prophage contig headers\n",
    "    sed -i -e \"s/\\s/__excised_start_/g\" -e \"s/-/_end_/g\" -e \"s/\\//_len_/g\" -e \"s/|/_/\" -e \"s/|//g\" 3.virus_prediction/6.perSample_checkv/S${i}_perSample_derep_checkv_out/cat_checkv_out.fasta\n",
    "done\n",
    "\n",
    "# Coassembly\n",
    " cat 3.virus_prediction/6.perSample_checkv/coassembly_perSample_derep_checkv_out/viruses.fna 3.virus_prediction/6.perSample_checkv/coassembly_perSample_derep_checkv_out/proviruses.fna > 3.virus_prediction/6.perSample_checkv/coassembly_perSample_derep_checkv_out/cat_checkv_out.fasta \n",
    "    # modify checkv prophage contig headers\n",
    "    sed -i -e \"s/\\s/__excised_start_/g\" -e \"s/-/_end_/g\" -e \"s/\\//_len_/g\" -e \"s/|/_/\" -e \"s/|//g\" 3.virus_prediction/6.perSample_checkv/coassembly_perSample_derep_checkv_out/cat_checkv_out.fasta\n",
    "\n",
    "# Mini assemblies\n",
    "for j in healthy diseased; do\n",
    "    cat 3.virus_prediction/6.perSample_checkv/${j}_coassembly_perSample_derep_checkv_out/viruses.fna 3.virus_prediction/6.perSample_checkv/${j}_coassembly_perSample_derep_checkv_out/proviruses.fna > 3.virus_prediction/6.perSample_checkv/${j}_coassembly_perSample_derep_checkv_out/cat_checkv_out.fasta \n",
    "    # modify checkv prophage contig headers\n",
    "    sed -i -e \"s/\\s/__excised_start_/g\" -e \"s/-/_end_/g\" -e \"s/\\//_len_/g\" -e \"s/|/_/\" -e \"s/|//g\" 3.virus_prediction/6.perSample_checkv/${j}_coassembly_perSample_derep_checkv_out/cat_checkv_out.fasta\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af0fa8b7",
   "metadata": {},
   "source": [
    "***Add checkv results to summary_table_VIRUSES.txt***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03c393b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd <working_directory>/WTS/\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load Python/3.8.2-gimkl-2020a\n",
    "python3\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Loop through all individual sample summary_tables, add key checkv results columns and write out.\n",
    "for i in range(1, 14):\n",
    "    # Import summary table (manually set dtypes)\n",
    "    summary_table = pd.read_csv('3.virus_prediction/4.summary_tables/S'+str(i)+'_summary_table_VIRUSES.txt', sep='\\t', \n",
    "                                dtype={'contig_ID': str, 'nt_contig_ID': str, 'nt_sacc': str, 'nt_stitle': str, 'nt_pident': float, \n",
    "                                       'nt_hit_length': float, 'nt_evalue': float, 'nt_Kingdom': str, \n",
    "                                       'nr_contig_ID': str, 'length': float, 'nr_stitle': str, 'nr_pident': float, \n",
    "                                       'nr_hit_length': float, 'nr_evalue': float, 'nr_Kingdom': str, \n",
    "                                       'rdrp_contig_ID': str, 'rdrp_stitle':str, 'rdrp_pident': float, \n",
    "                                       'rdrp_hit_length': float, 'rdrp_evalue': float, \n",
    "                                       'vibrant_contig_ID': str, 'vibrant_total_genes': float, \n",
    "                                       'vibrant_KEGG_genes': float, 'vibrant_KEGG_v_score': float, 'vibrant_Pfam_genes': float, \n",
    "                                       'vibrant_Pfam_v_score': float, 'vibrant_VOG_genes': float, 'vibrant_VOG_v_score': float, \n",
    "                                       'vsort2_contig_ID': str, 'vsort2_max_score': float, 'vsort2_max_score_group': str, \n",
    "                                       'vsort2_hallmark_genes': float, 'vsort2_viral_component': float, 'vsort2_cellular_component': float})        \n",
    "    # Import checkv summary results\n",
    "    checkv = pd.read_csv('3.virus_prediction/6.perSample_checkv/S'+str(i)+'_perSample_derep_checkv_out/quality_summary.tsv', sep='\\t', \n",
    "                        dtype={'contig_id': str, 'contig_length': float, 'provirus': str, 'proviral_length': float, \n",
    "                               'gene_count': float, 'viral_genes': float, 'host_genes': float, 'checkv_quality': str, \n",
    "                               'miuvig_quality': str, 'completeness': float, 'completeness_method': str, \n",
    "                               'contamination': float, 'kmer_freq': float, 'warnings': str})\n",
    "    checkv = checkv.add_prefix('checkv_')\n",
    "    checkv['contig_ID'] = checkv['checkv_contig_id']\n",
    "    # Merge with summary_table\n",
    "    summary_table = pd.merge(summary_table, checkv, left_on=\"contig_ID\", right_on=\"contig_ID\", how='outer')\n",
    "    # Output summary table\n",
    "    summary_table.to_csv('3.virus_prediction/4.summary_tables/S'+str(i)+'_summary_table_VIRUSES_checkv.txt', sep='\\t', index=False)\n",
    "\n",
    "\n",
    "\n",
    "# Coassembly\n",
    "summary_table = pd.read_csv('3.virus_prediction/4.summary_tables/coassembly_summary_table_VIRUSES.txt', sep='\\t', \n",
    "                            dtype={'contig_ID': str, 'nt_contig_ID': str, 'nt_sacc': str, 'nt_stitle': str, 'nt_pident': float, \n",
    "                                   'nt_hit_length': float, 'nt_evalue': float, 'nt_Kingdom': str, \n",
    "                                   'nr_contig_ID': str, 'length': float, 'nr_stitle': str, 'nr_pident': float, \n",
    "                                   'nr_hit_length': float, 'nr_evalue': float, 'nr_Kingdom': str, \n",
    "                                   'rdrp_contig_ID': str, 'rdrp_stitle':str, 'rdrp_pident': float, \n",
    "                                   'rdrp_hit_length': float, 'rdrp_evalue': float, \n",
    "                                   'vibrant_contig_ID': str, 'vibrant_total_genes': float, \n",
    "                                   'vibrant_KEGG_genes': float, 'vibrant_KEGG_v_score': float, 'vibrant_Pfam_genes': float, \n",
    "                                   'vibrant_Pfam_v_score': float, 'vibrant_VOG_genes': float, 'vibrant_VOG_v_score': float, \n",
    "                                   'vsort2_contig_ID': str, 'vsort2_max_score': float, 'vsort2_max_score_group': str, \n",
    "                                   'vsort2_hallmark_genes': float, 'vsort2_viral_component': float, 'vsort2_cellular_component': float})        \n",
    "\n",
    "# Import checkv summary results\n",
    "checkv = pd.read_csv('3.virus_prediction/6.perSample_checkv/coassembly_perSample_derep_checkv_out/quality_summary.tsv', sep='\\t', \n",
    "                    dtype={'contig_id': str, 'contig_length': float, 'provirus': str, 'proviral_length': float, \n",
    "                           'gene_count': float, 'viral_genes': float, 'host_genes': float, 'checkv_quality': str, \n",
    "                           'miuvig_quality': str, 'completeness': float, 'completeness_method': str, \n",
    "                           'contamination': float, 'kmer_freq': float, 'warnings': str})\n",
    "\n",
    "# Add prefix to checkv column headers\n",
    "checkv = checkv.add_prefix('checkv_')\n",
    "\n",
    "# Add 'contig_ID' column to checkv\n",
    "checkv['contig_ID'] = checkv['checkv_contig_id']\n",
    "\n",
    "# Merge with summary_table\n",
    "summary_table = pd.merge(summary_table, checkv, left_on=\"contig_ID\", right_on=\"contig_ID\", how='outer')\n",
    "\n",
    "# Output summary table\n",
    "summary_table.to_csv('3.virus_prediction/4.summary_tables/coassembly_summary_table_VIRUSES_checkv.txt', sep='\\t', index=False)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Mini assemblies\n",
    "for j in healthy diseased:\n",
    "    summary_table = pd.read_csv('3.virus_prediction/4.summary_tables/'+str(j)+'_coassembly_summary_table_VIRUSES.txt', sep='\\t', \n",
    "                                dtype={'contig_ID': str, 'nt_contig_ID': str, 'nt_sacc': str, 'nt_stitle': str, 'nt_pident': float, \n",
    "                                    'nt_hit_length': float, 'nt_evalue': float, 'nt_Kingdom': str, \n",
    "                                    'nr_contig_ID': str, 'length': float, 'nr_stitle': str, 'nr_pident': float, \n",
    "                                    'nr_hit_length': float, 'nr_evalue': float, 'nr_Kingdom': str, \n",
    "                                    'rdrp_contig_ID': str, 'rdrp_stitle':str, 'rdrp_pident': float, \n",
    "                                    'rdrp_hit_length': float, 'rdrp_evalue': float, \n",
    "                                    'vibrant_contig_ID': str, 'vibrant_total_genes': float, \n",
    "                                    'vibrant_KEGG_genes': float, 'vibrant_KEGG_v_score': float, 'vibrant_Pfam_genes': float, \n",
    "                                    'vibrant_Pfam_v_score': float, 'vibrant_VOG_genes': float, 'vibrant_VOG_v_score': float, \n",
    "                                    'vsort2_contig_ID': str, 'vsort2_max_score': float, 'vsort2_max_score_group': str, \n",
    "                                    'vsort2_hallmark_genes': float, 'vsort2_viral_component': float, 'vsort2_cellular_component': float})        \n",
    "\n",
    "    # Import checkv summary results\n",
    "    checkv = pd.read_csv('3.virus_prediction/6.perSample_checkv/'+str(j)+'_coassembly_perSample_derep_checkv_out/quality_summary.tsv', sep='\\t', \n",
    "                        dtype={'contig_id': str, 'contig_length': float, 'provirus': str, 'proviral_length': float, \n",
    "                            'gene_count': float, 'viral_genes': float, 'host_genes': float, 'checkv_quality': str, \n",
    "                            'miuvig_quality': str, 'completeness': float, 'completeness_method': str, \n",
    "                            'contamination': float, 'kmer_freq': float, 'warnings': str})\n",
    "\n",
    "    # Add prefix to checkv column headers\n",
    "    checkv = checkv.add_prefix('checkv_')\n",
    "\n",
    "    # Add 'contig_ID' column to checkv\n",
    "    checkv['contig_ID'] = checkv['checkv_contig_id']\n",
    "\n",
    "    # Merge with summary_table\n",
    "    summary_table = pd.merge(summary_table, checkv, left_on=\"contig_ID\", right_on=\"contig_ID\", how='outer')\n",
    "\n",
    "    # Output summary table\n",
    "    summary_table.to_csv('3.virus_prediction/4.summary_tables/'+str(j)+'_coassembly_summary_table_VIRUSES_checkv.txt', sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67f5cdf4",
   "metadata": {},
   "source": [
    "***OPTIONAL: Add sampleIDs to contig IDs***\n",
    "\n",
    "Add sampled IDs to all contig IDs so that each of the contigs that make up dereplicated contigs can be traced back to the samples they were derived from during downstream analyses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cda0e842",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd <working_directory>/WTS/\n",
    "\n",
    "for i in {1..13}; do\n",
    "    sampleID=$(echo S${i})\n",
    "    sed -i -e \"s/>/>${sampleID}_/g\" 3.virus_prediction/6.perSample_checkv/S${i}_perSample_derep_checkv_out/cat_checkv_out.fasta\n",
    "    done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65934a9f",
   "metadata": {},
   "source": [
    "Find retroviral contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74af2c3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd <working_directory>/WTS/\n",
    "\n",
    "for file in 3.virus_prediction/4.summary_tables/*_summary_table_VIRUSES_checkv.txt; do\n",
    "    echo ${file}\n",
    "    grep -n 'retro' ${file} >> retro.txt\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74d08fed",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52f0ec25",
   "metadata": {},
   "source": [
    "## Dereplication of contigs across all samples\n",
    "\n",
    "- Dereplicate contigs via Sullivan group's `Cluster_genomes_5.1.pl` approach (mummer) to generate 'viral OTUs' (vOTUs)\n",
    "    - From Roux et al. (2017) Benchmarking viromics: an in silico evaluation of metagenome-enabled estimates of viral community composition and diversity: \"Contigs from all samples were clustered with nucmer (Delcher, Salzberg & Phillippy, 2003) at >95% ANI across >80% of their lengths, as in (Brum et al., 2015; Gregory et al., 2016), to generate a pool of non-redundant population contigs.\"\n",
    "    - **Note that the above quote was for DNA viruses**\n",
    "- Requires Mummer v4.x. \n",
    "\n",
    "For more info on Cluster_genomes_5.1.pl see [here](https://github.com/simroux/ClusterGenomes)\n",
    "\n",
    "***Prep: download Cluster_genomees_5.1.pl***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22e1b81b",
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p <working_directory>/WTS/Software\n",
    "cd <working_directory>/WTS/Software\n",
    "\n",
    "# Download the script and set executable permissions\n",
    "wget https://raw.githubusercontent.com/simroux/ClusterGenomes/master/Cluster_genomes_5.1.pl\n",
    "chmod 755 <working_directory>/WTS/Software/Cluster_genomes_5.1.pl\n",
    "\n",
    "# Install autodie pearl module (dependency for Cluster_genomes.pl)\n",
    "cpan autodie"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21abb314",
   "metadata": {},
   "source": [
    "***Prep: Concatenate fasta files together for cluster_genomes.pl***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "218a15a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd <working_directory>/WTS/\n",
    "mkdir -p 4a.cluster_genomes/\n",
    "\n",
    "# Create output file\n",
    "> 4a.cluster_genomes/cat_AllSamples.fasta\n",
    "\n",
    "# Concatenate all samples into output file\n",
    "for i in {1..13}; do\n",
    "    cat 3.virus_prediction/6.perSample_checkv/S${i}_perSample_derep_checkv_out/cat_checkv_out.fasta >> 4a.cluster_genomes/cat_All_indv_Samples.fasta\n",
    "done\n",
    "\n",
    "for j in healthy diseased; do\n",
    "    cat 3.virus_prediction/6.perSample_checkv/${j}_coassembly_perSample_derep_checkv_out/cat_checkv_out.fasta >> 4a.cluster_genomes/cat_HD_Samples.fasta\n",
    "done\n",
    "\n",
    "cat 3.virus_prediction/6.perSample_checkv/coassembly_perSample_derep_checkv_out/cat_checkv_out.fasta 4a.cluster_genomes/cat_HD_Samples.fasta >> 4a.cluster_genomes/cat_All_coassemblies.fasta\n",
    "\n",
    "cat 4a.cluster_genomes/cat_All_coassemblies.fasta 4a.cluster_genomes/cat_All_indv_Samples.fasta >> 4a.cluster_genomes/cat_AllSamples.fasta\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7f74a50",
   "metadata": {},
   "source": [
    "***Optional: filter out short contigs (e.g. contigs < 500 kbp)***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "711b2368",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd <working_directory>/WTS/\n",
    "\n",
    "# load seqmagick\n",
    "module purge\n",
    "module load seqmagick/0.7.0-gimkl-2017a-Python-3.6.3\n",
    "\n",
    "# Filter out short contigs \n",
    "seqmagick convert --min-length 500 4a.cluster_genomes/cat_AllSamples.fasta 4a.cluster_genomes/cat_AllSamples.m500.fasta\n",
    "\n",
    "# Compare total number of contigs for each\n",
    "grep -c \">\" 4a.cluster_genomes/cat_AllSamples.fasta\n",
    "\n",
    "grep -c \">\" 4a.cluster_genomes/cat_AllSamples.m500.fasta\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd9e1aa7",
   "metadata": {},
   "source": [
    "***Run cluster_genomes.pl***\n",
    "\n",
    "Run at min identity = 95% similarity over at least 80% of the shortest contig\n",
    "\n",
    "`sbatch scripts/wts_5_derep_clusterGenomes.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca24207f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_5_derep_clusterGenomes\n",
    "#SBATCH --time 00:10:00\n",
    "#SBATCH --mem=4GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=32\n",
    "#SBATCH -e slurm_output/wts_5_derep_clusterGenomes_%j.err\n",
    "#SBATCH -o slurm_output/wts_5_derep_clusterGenomes_%j.out\n",
    "\n",
    "# Set up working directories\n",
    "cd <working_directory>/WTS/4a.cluster_genomes\n",
    "\n",
    "# Set variables\n",
    "module purge\n",
    "cluster_genomes_path='/nesi/nobackup/uoa03068/Annie/RNA_seq/Software'\n",
    "mummer_path='/nesi/project/uoa02469/Software/mummer_v4.0.0/bin/'\n",
    "\n",
    "# Run for all contigs\n",
    "${cluster_genomes_path}/Cluster_genomes_5.1.pl \\\n",
    "-f cat_AllSamples.fasta \\\n",
    "-d ${mummer_path} \\\n",
    "-t 32 \\\n",
    "-c 80 \\\n",
    "-i 95"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e6cb46bd",
   "metadata": {},
   "source": [
    "***Check total number of clustered contigs (vOTUs) remaining***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af2360c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd <working_directory>/WTS/4a.cluster_genomes\n",
    "\n",
    "grep -c \">\" cat_AllSamples_95-80.fna"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7856a1c8",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4edf0835",
   "metadata": {},
   "source": [
    "## Modify derep contig headers to be `vOTU_n`\n",
    "\n",
    "- CheckV downstream fails with contig headers longer than 10,000 characters. The previous dereplication steps can result in some headers being too long and checkv failing.\n",
    "- Replace all headers with `vOTU_n` and create a lookup table file of `vOTU_n` IDs against full contig headers (for reference later).\n",
    "\n",
    "**Python script to do the above:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41a20943",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd <working_directory>/WTS/4a.cluster_genomes\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load Python/3.8.2-gimkl-2020a\n",
    "python3\n",
    "\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser\n",
    "\n",
    "fasta_in = 'cat_AllSamples_95-80.fna'\n",
    "fasta_out = 'vOTUs.fa'\n",
    "lookup_table_out = 'vOTUs_virID_lookup_table.txt'\n",
    "\n",
    "# Read in fasta file, looping through each contig\n",
    "# rename contig headers with incrementing vOTU_n headers\n",
    "# write out new vOTUs.fa file and tab-delimited table file of matching vOTU_n and contigID headers.\n",
    "i=1\n",
    "with open(fasta_in, 'r') as read_fasta:\n",
    "    with open(fasta_out, 'w') as write_fasta:\n",
    "        with open (lookup_table_out, 'w') as write_table:\n",
    "            write_table.write(\"vOTU_ID\" + \"\\t\" + \"cluster_contigID\" + \"\\n\")\n",
    "            for name, seq in SimpleFastaParser(read_fasta):\n",
    "                write_table.write(\"vOTU_\" + str(i) + \"\\t\" + name + \"\\n\")\n",
    "                write_fasta.write(\">\" + \"vOTU_\" + str(i) + \"\\n\" + str(seq) + \"\\n\")\n",
    "                i += 1\n",
    "\n",
    "quit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4be8386",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "888bd1ed",
   "metadata": {},
   "source": [
    "**Downstream vOTU processing:**\n",
    "\n",
    "Re-run vOTUs through the blast searches, checkv, and generate new summary table (seemed like the easiest way to get the blast search info back in the mix after generating the clustered vOTU contig set)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e241e672",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab1cef1a",
   "metadata": {},
   "source": [
    "## vOTUs: 3 x blast searches\n",
    "\n",
    "Re-run 3 x blast searches (nr, nt, rdrp) for the dereplicated viral contig sets (vOTUs).\n",
    "\n",
    "***RdRP blastx search: vOTUs***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c95e4832",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up working directory\n",
    "cd <working_directory>/WTS/4a.cluster_genomes\n",
    "mkdir -p blast\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load DIAMOND/0.9.32-GCC-9.2.0\n",
    "\n",
    "# Database variable\n",
    "search_db='/nesi/project/uoa02469/Databases/viral_rdrp_protein_database/viral_rdrp.dmnd'\n",
    "\n",
    "# Run diamond blastx search\n",
    "diamond blastx \\\n",
    "-q vOTUs.fa \\\n",
    "-d ${search_db} \\\n",
    "-o blast/vOTUs_RdRp.txt \\\n",
    "-e 1E-5 -k 3 -p 6 -f 6 qseqid qlen sseqid stitle pident length evalue --more-sensitive"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a1e1a98",
   "metadata": {},
   "source": [
    "***nr blastx search: vOTUs***\n",
    "\n",
    "`sbatch scripts/wts_5_vOTUs_blast_nr.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "480f1466",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_5_vOTUs_blast_nr\n",
    "#SBATCH --time 07:00:00\n",
    "#SBATCH --mem=20GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=24\n",
    "#SBATCH -e slurm_output/wts_5_vOTUs_blast_nr_%j.err\n",
    "#SBATCH -o slurm_output/wts_5_vOTUs_blast_nr_%j.out\n",
    "\n",
    "# Change to working directory\n",
    "cd <working_directory>/WTS/4a.cluster_genomes\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load DIAMOND/2.0.6-GCC-9.2.0\n",
    "\n",
    "# Database variable\n",
    "search_db='/nesi/project/uoa02469/Databases/NCBI_20210823_dmnd/nr_wTax_FULL.dmnd'\n",
    "\n",
    "# Run diamond blastx search\n",
    "srun diamond blastx \\\n",
    "-q vOTUs.fa \\\n",
    "-d ${search_db} \\\n",
    "-o blast/vOTUs_nr.txt \\\n",
    "-e 1E-5 -k 3 -p 24 -f 6 qseqid qlen sseqid stitle pident length evalue sskingdoms --more-sensitive"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea2b586c",
   "metadata": {},
   "source": [
    "***nt blastn search: vOTUs¶***\n",
    "\n",
    "`sbatch scripts/wts_5_vOTUs_blast_nt.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "31a1c024",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_5_vOTUs_blast_nt\n",
    "#SBATCH --time 01:00:00\n",
    "#SBATCH --mem=4GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=24\n",
    "#SBATCH -e slurm_output/wts_5_vOTUs_blast_nt_%j.err\n",
    "#SBATCH -o slurm_output/wts_5_vOTUs_blast_nt_%j.out\n",
    "\n",
    "# Change to working directory\n",
    "cd <working_directory>/WTS/4a.cluster_genomes\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load BLAST/2.10.0-GCC-9.2.0\n",
    "module load BLASTDB/2021-01\n",
    "\n",
    "# Database variable\n",
    "search_db=nt\n",
    "\n",
    "# Run diamond blastx search\n",
    "blastn \\\n",
    "-query vOTUs.fa \\\n",
    "-db ${search_db} \\\n",
    "-out blast/vOTUs_nt.txt \\\n",
    "-max_target_seqs 5 -num_threads 24 -evalue 1E-10 \\\n",
    "-outfmt \"6 qseqid sacc salltitles pident length evalue sskingdoms\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "284ad876",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef32c2c5",
   "metadata": {},
   "source": [
    "## vOTUs: checkV\n",
    "\n",
    "Re-run CheckV on vOTUs to output checkv stats on the final clustered/dereplicated contig set.\n",
    "\n",
    "`sbatch scripts/wts_5_vOTUs_checkv.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a199c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash -e\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_5_vOTUs_checkv\n",
    "#SBATCH --time 00:10:00\n",
    "#SBATCH --mem=4GB\n",
    "#SBATCH --ntasks=1\n",
    "#SBATCH --cpus-per-task=16\n",
    "#SBATCH -e slurm_output/wts_5_vOTUs_checkv_%j.err\n",
    "#SBATCH -o slurm_output/wts_5_vOTUs_checkv_%j.out\n",
    "\n",
    "# Set up working directories\n",
    "cd <working_directory>/WTS/\n",
    "mkdir -p 4b.checkv_analyses_vOTUs\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load CheckV/0.7.0-gimkl-2020a-Python-3.8.2\n",
    "\n",
    "# Run main analyses \n",
    "checkv_in=\"4a.cluster_genomes/vOTUs.fa\"\n",
    "checkv_out=\"4b.checkv_analyses_vOTUs/checkv_out_vOTUs\"\n",
    "srun checkv end_to_end ${checkv_in} ${checkv_out} -t 16 --quiet"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3198a6c0",
   "metadata": {},
   "source": [
    "**CheckV: Concatenate output fasta files (viruses.fna and proviruses.fna)¶**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0d5e65b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd <working_directory>/WTS/\n",
    "\n",
    "# Concatenate viruses and prophage files\n",
    "cat 4b.checkv_analyses_vOTUs/checkv_out_vOTUs/viruses.fna  4b.checkv_analyses_vOTUs/checkv_out_vOTUs/proviruses.fna >  4b.checkv_analyses_vOTUs/cat_checkv_out_vOTUs.fasta \n",
    "# Modify checkv prophage contig headers\n",
    "sed -i -e \"s/\\s/__excised_start_/g\" -e \"s/-/_end_/g\" -e \"s/\\//_len_/g\" -e \"s/|/_/\" -e \"s/|//g\" 4b.checkv_analyses_vOTUs/cat_checkv_out_vOTUs.fasta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e392e69",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9692e2a",
   "metadata": {},
   "source": [
    "## Optional: Further filtering of clustered contigs (vOTU)\n",
    "\n",
    "Filter (losely) based on Sullivan/Roux groups' curation thresholds.\n",
    "\n",
    "Based on info on [this protocols page](https://www.protocols.io/view/viral-sequence-identification-sop-with-virsorter2-5qpvoyqebg4o/v3?step=4) (as of 24 Aug 2021)\n",
    "\n",
    "- n.b. omit inclusion of VirSorter2 score and hallmark genes thresholds for filtering\n",
    "- keep any with: (viral gene > 0 OR (viral gene == 0 AND host gene == 0))\n",
    "- Downstream (e.g. in R) add manual checks for annotations of concern from [this list](https://bitbucket.org/MAVERICLab/virsorter2-sop/raw/03b8f28bee979e2b7fd99d7375d915c29c938339/resource/suspicious-gene.list), based on vibrant and/or DRAM-v annotations.\n",
    "- Downstream, you can also choose whether to further quality filter (e.g. to remove \"short\" contigs, or those categorised by checkv as 'Low-quality')\n",
    "\n",
    "**checkv_filter_contigs.py: run on checkv output from vOTUs**\n",
    "\n",
    "Outputs:\n",
    "\n",
    "- `..._filtered.fna` : filtered fasta file\n",
    "- `..._filtered_quality_summary.tsv` : filtered checkv quality_summary output file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26213f17",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load module(s)\n",
    "module purge\n",
    "module load Python/3.8.2-gimkl-2020a\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/\n",
    "script_path='/nesi/project/uoa02469/custom-scripts/MikeH/'\n",
    "\n",
    "# Run for vOTUs\n",
    "${script_path}/checkv_filter_contigs.py \\\n",
    "    --checkv_dir_input 4b.checkv_analyses_vOTUs/checkv_out_vOTUs/ \\\n",
    "    --output_prefix 4b.checkv_analyses_vOTUs/vOTUs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddcaf5bd",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34a4fe58",
   "metadata": {},
   "source": [
    "## WTS vOTUs read mapping: vOTU read coverage\n",
    "\n",
    "- Optional: combine with dereplicated **WGS** contigs for read mapping to ensure reads aren't erroneously mapped to genes in viral genomes that are similar to transcripts from hosts (e.g. AMGs).\n",
    "    - In this case, ignore the counts for the wgs contigs. This is simply to ensure the transcript reads aren't erroneously mapped to putative RNA viruses. (Transcription read mapping for WGS contigs is included in the WGS data processing workflow)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "acc8ca39",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd <working_directory>/WTS/\n",
    "mkdir -p 5.wts_read_mapping_vOTUs\n",
    "\n",
    "# Concatenate WGS dereplicated contigs with contigs of putative RNA viruses\n",
    "cat <working_directory>/WGS/2.assembly_spades/dedupe/dereplicated_contigs.fna <working_directory>/WTS/4b.checkv_analyses_vOTUs/vOTUs_filtered.fna > 5.wts_read_mapping_vOTUs/cat_RNAviralContigs_and_DNAderepContigs.fna"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77a76984",
   "metadata": {},
   "source": [
    "***WTS vOTUs: BBMap, build reference index***\n",
    "\n",
    "`sbatch scripts/wts_6_read_mapping_vOTUs_index.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b663d22e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -A uoa03068\n",
    "#SBATCH -J wts_6_read_mapping_vOTUs_index\n",
    "#SBATCH --time 00:10:00\n",
    "#SBATCH --mem 4GB\n",
    "#SBATCH --ntasks 1\n",
    "#SBATCH --cpus-per-task 24\n",
    "#SBATCH -e slurm_output/wts_6_read_mapping_vOTUs_index_%j.err\n",
    "#SBATCH -o slurm_output/wts_6_read_mapping_vOTUs_index_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/5.wts_read_mapping_vOTUs/\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load BBMap/38.90-gimkl-2020a\n",
    "\n",
    "# Build index\n",
    "bbmap.sh -Xmx4g ref=cat_RNAviralContigs_and_DNAderepContigs.fna"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d2e9416",
   "metadata": {},
   "source": [
    "***WTS vOTUs: per-sample WTS read mapping, slurm array***\n",
    "\n",
    "`sbatch scripts/wts_6_read_mapping_vOTUs.sl`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5e58d01",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "#SBATCH -A your_project_account\n",
    "#SBATCH -J wts_6_read_mapping_vOTUs\n",
    "#SBATCH --time 01:00:00\n",
    "#SBATCH --mem 15GB\n",
    "#SBATCH --ntasks 1\n",
    "#SBATCH --array=1-13\n",
    "#SBATCH --cpus-per-task 30\n",
    "#SBATCH -e slurm_output/wts_6_read_mapping_vOTUs_%a_%j.err\n",
    "#SBATCH -o slurm_output/wts_6_read_mapping_vOTUs_%a_%j.out\n",
    "\n",
    "# Set up working directory\n",
    "cd <working_directory>/WTS/5.wts_read_mapping_vOTUs/\n",
    "mkdir -p wts/\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load BBMap/38.90-gimkl-2020a\n",
    "module load SAMtools/1.10-GCC-9.2.0\n",
    "\n",
    "# in/out varibles\n",
    "WTS_READ_DIR='<working_directory>/WTS/1.host_filtered'\n",
    "OUTPATH='wts'\n",
    "\n",
    "# Run read mapping\n",
    "srun bbmap.sh \\\n",
    "t=30 -Xmx15g ambiguous=best minid=0.95 \\\n",
    "in1=${WGS_READ_DIR}/S${SLURM_ARRAY_TASK_ID}_R1_hostFilt.fastq \\\n",
    "in2=${WGS_READ_DIR}/S${SLURM_ARRAY_TASK_ID}_R2_hostFilt.fastq \\\n",
    "covstats=${OUTPATH}/S${SLURM_ARRAY_TASK_ID}.wts.covstats.txt \\\n",
    "statsfile=${OUTPATH}/S${SLURM_ARRAY_TASK_ID}.wts.statsfile.txt \\\n",
    "out=${OUTPATH}/S${SLURM_ARRAY_TASK_ID}.wts.sam\n",
    "\n",
    "# Convert sam to bam\n",
    "samtools sort -@ 10 -o ${OUTPATH}/S${SLURM_ARRAY_TASK_ID}.wts.bam ${OUTPATH}/S${SLURM_ARRAY_TASK_ID}.wts.sam\n",
    "\n",
    "# Run pileup to extract read counts\n",
    "pileup.sh \\\n",
    "in=${OUTPATH}/S${SLURM_ARRAY_TASK_ID}.wts.sam \\\n",
    "rpkm=${OUTPATH}/S${SLURM_ARRAY_TASK_ID}.wts.covstats_pileup.txt "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76dfc50f",
   "metadata": {},
   "source": [
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6169f63a",
   "metadata": {},
   "source": [
    "## WTS vOTUs: Per-sample coverage calculations\n",
    "\n",
    "Per-sample coverages can be calculated at:\n",
    "\n",
    "- genome level (e.g. across multiple binned contigs)\n",
    "- contig level (e.g. for viral genomes represented by a single contig)\n",
    "- gene level\n",
    "\n",
    "`summarise_counts.py` (and associated script `summarise_counts.R`) generated by Michael Hoggard can output each of these sets of summaries depending on the input data (pileup or featurecounts) and output options.\n",
    "\n",
    "As we're interested in RNA viruses here, we will generate contig-level coverage summaries based on the output from *pileup*.\n",
    "\n",
    "Below, contig-level coverage summary outputs will be generated for:\n",
    "\n",
    "- WTS read mapping (indicative of *RNA* data WTS-contig read-coverages)\n",
    "\n",
    "**Preamble: summarise_counts.py arguments**\n",
    "\n",
    "Required\n",
    "\n",
    "- `--input (-i) counts_files.txt`: Input file(s) of counts data returned by featurecounts (gene_counts file) or pileup.sh (rpkm output files passed using wildcard).\n",
    "    - Note: for pileup.sh output files, this takes a wildcard to capture all the relevant input files (e.g. `--input *rpkm.txt`)\n",
    "- `--format (-f) ['featurecounts', 'pileup']` : Format of the input counts data. Must be 'featurecounts' or 'pileup'\n",
    "\n",
    "\n",
    "Optional\n",
    "- `--sample_mapping_file (-m) sample_mapping_file.txt` : Mapping file of unique sample identifier strings matching to sample groups (and library size (optional)).\n",
    "    - Columns must be named as: `sampleID`, `group`, `lib.size` (lib.size is an optional column).\n",
    "    - sampleID strings must also be present in the input files (in the counts column headers in output featurecounts file, or in the filenames of pileup.sh output files)\n",
    "- `--lib_norm (-n) ['total', 'mapped']` : Set whether to (re)calculate RPKM and FPKM based on total library size or mapped reads per sample. Generally (often based on single organism transcription analyses) total mapped reads is used. However, in some instances, such as metatranscriptome data from mixed organisms, you may wish to override this with total library size (total read counts in fastq files used for read mapping). (Default = 'mapped')\n",
    "- `--count_threshold (-t) numeric_threshold_value` : Set threshold to zero out low count values below the threshold. (Default = 1)\n",
    "- `--read_counts (-r) summary_read_counts.tsv` : Output file name for table of read counts (per-sample total library, mapped read, and filtered mapped read (based on count_threshold) counts). (Default = 'summary_read_counts.tsv')\n",
    "- `--edger_out (-e) summary_edgeR_glmQLFTest.tsv` : Output file name for summary table of EdgeR analysis of \"differentially expressed genes\" (DEG) across sample groups\n",
    "- `--output (-o) normalised_summary_count_table.tsv` : Output file name for summary count table (default = 'normalised_summary_count_table.tsv')\n",
    "\n",
    "\n",
    "**Required prep: generate sample mapping files**\n",
    "\n",
    "A (tab-delimited) sample mapping file is required if you wish to include TMM normalisation (as this is generated via EdgeR and requires sample grouping information).\n",
    "\n",
    "Note:\n",
    "\n",
    "- For read mapping WTS data to putative RNA viruses here, you will want to include the library size of the filtered **WTS** read files (i.e. those in `/working/dir/1.host_filtered`).\n",
    "- You can count the number of reads in each via `grep -c '@' ${file}`, and sum the R1 and R2 reads together\n",
    "- Required columns (spelling of column headers matters)\n",
    "    - `sampleID`: unique strings that identify each sample (these unique strings must also be present in the counts files file names)\n",
    "    - `group`: treatment groups for each sample (e.g. 'treatment', 'control'; or 'groupA', 'groupB', 'groupC' etc.)\n",
    "- Optional column\n",
    "    - `lib.size`: total library size per sample (total read counts in the quality trimmed and filtered fastq files used for read mapping). (NOTE: this column is required if `--format featurecounts` and `--lib_norm total` (for `--format pileup`, this data is present in the input files output by pileup's `rpkm=...` option))\n",
    "\n",
    "Example mapping file:\n",
    "\n",
    "| SampleID | group | lib.size | \n",
    "| :- | -: | -: | \n",
    "| Sample_1 | A | 1054234 | \n",
    "|Sample_2|A|1543619|\n",
    "|Sample_3|B|1246519|\n",
    "|Sample_4|B|1349855|\n",
    "|Sample_5|C|1644875|\n",
    "|Sample_6|C|1422537|\n",
    "\n",
    "Save this sample mapping file as, e.g:\n",
    "\n",
    "`working/dir/5.read_mapping/wts/wts_sample_mapping_file.tsv`\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5399deb",
   "metadata": {},
   "source": [
    "***WTS vOTUs: summarise_counts.py***\n",
    "\n",
    "\n",
    "| sampleID | group | lib.size |\n",
    "| :- | -: | -: | \n",
    "|S1|    A|  16260228|\n",
    "|S2|    A|   949022|\n",
    "|S3|    A|   8190158|\n",
    "|S4|    A|   628904|\n",
    "|S5|    A|   1480372|\n",
    "|S6|    A|   852728|\n",
    "|S7|    A|   3638000|\n",
    "|S8|    B|   5801764|\n",
    "|S9|    B|   5119314|\n",
    "|S10|   A|   664664|\n",
    "|S11|   A|   1485674|\n",
    "|S12|   B|   889510|\n",
    "|S13|   B|   570700|"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12f8e166",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd <working_directory>/WTS/\n",
    "\n",
    "# Load module(s)\n",
    "module purge\n",
    "module load Python/3.8.2-gimkl-2020a\n",
    "module load R/3.6.2-gimkl-2020a\n",
    "\n",
    "# Add location of summarise_counts.py and summarise_counts.R to PATH\n",
    "export PATH=\"/nesi/project/uoa02469/custom-scripts/MikeH:$PATH\"\n",
    "\n",
    "# Run summarise_counts.py\n",
    "summarise_counts.py \\\n",
    "    --input '5.wts_read_mapping_vOTUs/wts/*.covstats_pileup.txt' --format pileup \\\n",
    "    --sample_map 5.wts_read_mapping_vOTUs/wts/sample_mapping_file.txt \\\n",
    "    --lib_norm total \\\n",
    "    --count_threshold 10 \\\n",
    "    --read_counts 5.wts_read_mapping_vOTUs/wts/wts_summary_read_counts_RNA_vOTUs.tsv \\\n",
    "    --output 5.wts_read_mapping_vOTUs/wts/wts_summary_count_table_RNA_vOTUs.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "095eaf45",
   "metadata": {},
   "source": [
    "***"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
